{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.7.4"
    },
    "colab": {
      "name": "wormcells-viz-pipeline-example.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/WormBase/wormcells-notebooks/blob/main/wormcells_viz_pipeline_example.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "99lhIpQFfC5n"
      },
      "source": [
        "### This notebook will create a trained scVI model and write the data that is used by wormcells-viz\n",
        "\n",
        "It assumes that data has been wrangled into the WormBase standard anndata format:\n",
        "https://github.com/WormBase/anndata-wrangling\n",
        "\n",
        "The data processed here is downloaded from Caltech data, but any h5ad file following the WormBase standard anndata convention will work.\n",
        "\n",
        "Three separate anndata files (.h5ad) will be created: \n",
        "\n",
        "## For the expression heatmap\n",
        "This data is a 2D matrix of shape:\n",
        "$ x_{celltypes} \\times y_{genes} = x_{obs} \\times y_{var}  $ \n",
        "\n",
        "```\n",
        "anndata.obs = cell_types\n",
        "anndata.var = gene_id\n",
        "anndata.X = log10 scvi expression frequency values in the X field\n",
        "```\n",
        "![image](https://user-images.githubusercontent.com/12504176/120026581-21a77b80-bfa7-11eb-81db-39040cd76148.png)\n",
        "\n",
        "## For the gene histogram\n",
        "\n",
        "This data is a 3D tensor of shape: \n",
        "\n",
        "$ n_{celltypes} \\times n_{bins} \\times n_{genes} = n_{obs} \\times n_{var} \\times n_{layers} $  \n",
        "\n",
        "The anndata obs contains the cell types and var contains the histogram bins, the genes are stored in layers with the keys being the gene ID.\n",
        "\n",
        "We store the genes in the layers because each view in the wormcells-viz app show the histograms for a single gene, so this makes accessing the data simpler\n",
        "\n",
        "The histogram bin counts are computed from the scvi normalized expression values, binned in 100 bins from 10^-9 to 10^0\n",
        "\n",
        "```\n",
        "anndata.obs = cell_types\n",
        "anndata.var = bins of counts seen in range (100 evenly spaced bins, from -10 to 0) because the data is log10 transformed\n",
        "anndata.X = NOTHING (filled with vector of zeroes)\n",
        "anndata.layers[cell_type] = the key is the corresponding cell_type, each layer\n",
        "contains counts in each bin for all cell types\n",
        "anndata.uns['about'] = information about the dataset\n",
        "```\n",
        "![image](https://user-images.githubusercontent.com/12504176/120026750-67fcda80-bfa7-11eb-926d-74c2937f9a6b.png)\n",
        "\n",
        "## For the swarm plots\n",
        "\n",
        "This data is a 3D tensor of shape: \n",
        "\n",
        "$ x_{celltypes} \\times y_{genes} \\times z_{celltypes} \\times n_{genes} = x_{obs} \\times y_{var} \\times z_{layers} $ \n",
        "\n",
        "Notice that the cell types are repeated along two dimensions, because this data contains the results of pairwise DE comparisons among each cell type in the data.\n",
        "\n",
        "**Plus** a 2D matrix of shape:\n",
        "$ x_{celltypes} \\times y_{genes} = x_{obs} \\times y_{var}  $\n",
        "\n",
        "Because `anndata.uns[celltype]` constains a dataframe with global differnetial expression results for that celltype.\n",
        "\n",
        "Finally, `anndata.uns['heatmap']` contains the 2D matrix with log10 scvi expression frequency heatmap data, with genes in the index and cell types in the columns. This can be used to display the expression of each tissue upon mouseover.\n",
        "\n",
        "```\n",
        "anndata.obs = cell_types\n",
        "anndata.var = gene_id\n",
        "anndata.X = NOTHING (filled with vector of zeroes)\n",
        "anndata.layers[cell_type] = mean log fold change for a given cell type for all genes \n",
        "anndata.uns[cell_type] = contain the DE result of the corresponding cell type vs all other cells this can be used for ordering the genes by p-value, expression, and by log fold change lfc max/min/median/mean/std\n",
        "\n",
        "anndata.uns['heatmap']=dataframe with genes in index and cell types in columns containing the log10 of the scvi expression frequency for each cell type \n",
        "anndata.uns['about'] = information about the dataset\n",
        "```\n",
        "\n",
        "![image](https://user-images.githubusercontent.com/12504176/120026814-7a771400-bfa7-11eb-8c0d-5c3efbccd722.png)\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BIyPQuw1fFZx",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "dd328782-80a2-45b0-c355-5f396e8edd60"
      },
      "source": [
        "!pip install scvi-tools==0.11.0 --quiet"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\u001b[?25l\r\u001b[K     |█▋                              | 10kB 26.5MB/s eta 0:00:01\r\u001b[K     |███▎                            | 20kB 21.2MB/s eta 0:00:01\r\u001b[K     |█████                           | 30kB 16.8MB/s eta 0:00:01\r\u001b[K     |██████▋                         | 40kB 15.1MB/s eta 0:00:01\r\u001b[K     |████████▎                       | 51kB 8.6MB/s eta 0:00:01\r\u001b[K     |█████████▉                      | 61kB 9.9MB/s eta 0:00:01\r\u001b[K     |███████████▌                    | 71kB 9.4MB/s eta 0:00:01\r\u001b[K     |█████████████▏                  | 81kB 10.3MB/s eta 0:00:01\r\u001b[K     |██████████████▉                 | 92kB 10.0MB/s eta 0:00:01\r\u001b[K     |████████████████▌               | 102kB 8.2MB/s eta 0:00:01\r\u001b[K     |██████████████████              | 112kB 8.2MB/s eta 0:00:01\r\u001b[K     |███████████████████▊            | 122kB 8.2MB/s eta 0:00:01\r\u001b[K     |█████████████████████▍          | 133kB 8.2MB/s eta 0:00:01\r\u001b[K     |███████████████████████         | 143kB 8.2MB/s eta 0:00:01\r\u001b[K     |████████████████████████▊       | 153kB 8.2MB/s eta 0:00:01\r\u001b[K     |██████████████████████████▎     | 163kB 8.2MB/s eta 0:00:01\r\u001b[K     |████████████████████████████    | 174kB 8.2MB/s eta 0:00:01\r\u001b[K     |█████████████████████████████▋  | 184kB 8.2MB/s eta 0:00:01\r\u001b[K     |███████████████████████████████▎| 194kB 8.2MB/s eta 0:00:01\r\u001b[K     |████████████████████████████████| 204kB 8.2MB/s \n",
            "\u001b[K     |████████████████████████████████| 245kB 13.8MB/s \n",
            "\u001b[K     |████████████████████████████████| 215kB 14.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 81kB 9.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 634kB 16.0MB/s \n",
            "\u001b[K     |████████████████████████████████| 133kB 24.3MB/s \n",
            "\u001b[K     |████████████████████████████████| 808kB 22.7MB/s \n",
            "\u001b[K     |████████████████████████████████| 51kB 7.8MB/s \n",
            "\u001b[K     |████████████████████████████████| 112kB 48.6MB/s \n",
            "\u001b[K     |████████████████████████████████| 829kB 39.1MB/s \n",
            "\u001b[K     |████████████████████████████████| 10.6MB 47.8MB/s \n",
            "\u001b[K     |████████████████████████████████| 276kB 49.2MB/s \n",
            "\u001b[K     |████████████████████████████████| 645kB 46.1MB/s \n",
            "\u001b[K     |████████████████████████████████| 1.3MB 43.6MB/s \n",
            "\u001b[K     |████████████████████████████████| 143kB 55.5MB/s \n",
            "\u001b[K     |████████████████████████████████| 296kB 58.6MB/s \n",
            "\u001b[?25h  Building wheel for future (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[31mERROR: tensorflow 2.5.0 has requirement tensorboard~=2.5, but you'll have tensorboard 2.4.1 which is incompatible.\u001b[0m\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 90
        },
        "id": "OqXVTAaBfC5o",
        "outputId": "99638da6-3e35-4ce0-8e85-9ea8d704a996"
      },
      "source": [
        "import anndata\n",
        "import scvi\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import itertools\n",
        "from tqdm import tqdm\n",
        "import pandas as pd\n",
        "scvi.__version__\n"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/numba/np/ufunc/parallel.py:363: NumbaWarning: \u001b[1mThe TBB threading layer requires TBB version 2019.5 or later i.e., TBB_INTERFACE_VERSION >= 11005. Found TBB_INTERFACE_VERSION = 9107. The TBB threading layer is disabled.\u001b[0m\n",
            "  warnings.warn(problem)\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'0.11.0'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 2
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5Vp5bnMefC5p",
        "outputId": "7f60e94b-0126-4002-8f10-4afc6d67b359"
      },
      "source": [
        "!wget -O bendavid2021.h5ad https://data.caltech.edu/tindfiles/serve/a90cc35d-903c-494a-a06c-833e138688fc/     "
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-06-10 05:52:57--  https://data.caltech.edu/tindfiles/serve/a90cc35d-903c-494a-a06c-833e138688fc/\n",
            "Resolving data.caltech.edu (data.caltech.edu)... 34.250.186.131, 34.251.243.51, 34.252.192.177\n",
            "Connecting to data.caltech.edu (data.caltech.edu)|34.250.186.131|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 FOUND\n",
            "Location: https://tind-caltechdata.s3.amazonaws.com:443/127be439c20d4d16959cd5caea1d005f?Signature=JJgB3tOsVmWhzibH7R2RdOHW98U%3D&Expires=1623304677&AWSAccessKeyId=AKIAJSK7IFF5HDDM5UBQ&response-content-disposition=attachment%3Bfilename%3D%22bendavid2021.h5ad%22&response-content-type=application/force-download [following]\n",
            "--2021-06-10 05:52:57--  https://tind-caltechdata.s3.amazonaws.com/127be439c20d4d16959cd5caea1d005f?Signature=JJgB3tOsVmWhzibH7R2RdOHW98U%3D&Expires=1623304677&AWSAccessKeyId=AKIAJSK7IFF5HDDM5UBQ&response-content-disposition=attachment%3Bfilename%3D%22bendavid2021.h5ad%22&response-content-type=application/force-download\n",
            "Resolving tind-caltechdata.s3.amazonaws.com (tind-caltechdata.s3.amazonaws.com)... 52.218.30.234\n",
            "Connecting to tind-caltechdata.s3.amazonaws.com (tind-caltechdata.s3.amazonaws.com)|52.218.30.234|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 155794162 (149M) [application/force-download]\n",
            "Saving to: ‘bendavid2021.h5ad’\n",
            "\n",
            "bendavid2021.h5ad   100%[===================>] 148.58M  19.8MB/s    in 8.7s    \n",
            "\n",
            "2021-06-10 05:53:07 (17.1 MB/s) - ‘bendavid2021.h5ad’ saved [155794162/155794162]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IxWIBInFfC5q"
      },
      "source": [
        "adata=anndata.read_h5ad('bendavid2021.h5ad')"
      ],
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Q-eWurd5fC5q",
        "outputId": "2403e494-0c98-4fb0-bf84-f2e977a7d574"
      },
      "source": [
        "adata"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "AnnData object with n_obs × n_vars = 55508 × 20138\n",
              "    obs: 'study', 'sample_batch', 'sample', 'cell_type', 'cell_subtype', 'sample_description', 'barcode'\n",
              "    var: 'gene_id', 'gene_name', 'wbps_transcript_id', 'chromosome_name', 'start_position', 'end_position', 'strand', 'external_gene_id', 'external_transcript_id', 'wormbase_locus', 'wormbase_gseq'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 352
        },
        "id": "NClNic3gfC5r",
        "outputId": "6eb1c503-07bc-4099-8601-cac280330572"
      },
      "source": [
        "adata.obs.T"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>F4_1+TGTAACGGTTAGCTAC-1</th>\n",
              "      <th>F4_1+GGCAGTCCAGCCTATA-1</th>\n",
              "      <th>F4_1+AAGTACCGTCATCCCT-1</th>\n",
              "      <th>F4_1+AAGATAGTCCCTCTAG-1</th>\n",
              "      <th>F4_1+ACCAAACCAGCTGTAT-1</th>\n",
              "      <th>F4_1+CCGTTCATCACTCACC-1</th>\n",
              "      <th>F4_1+ATGTCCCCAATCGCCG-1</th>\n",
              "      <th>F4_1+GATTCGATCGTCAACA-1</th>\n",
              "      <th>F4_1+TACGTCCAGTTGCTGT-1</th>\n",
              "      <th>F4_1+GACCGTGTCCGTGCGA-1</th>\n",
              "      <th>F4_1+CCGATGGTCTTACCAT-1</th>\n",
              "      <th>F4_1+TCTGTCGTCTCAATCT-1</th>\n",
              "      <th>F4_1+GACCGTGAGCGTGTTT-1</th>\n",
              "      <th>F4_1+AGATGCTGTAGCTGCC-1</th>\n",
              "      <th>F4_1+ACACCAACAGACGGAT-1</th>\n",
              "      <th>F4_1+GCGGATCTCCCATGGG-1</th>\n",
              "      <th>F4_1+TACCGAAAGCTGTGCC-1</th>\n",
              "      <th>F4_1+AGCTTCCCAGATTAAG-1</th>\n",
              "      <th>F4_1+CATTGCCAGGTACAAT-1</th>\n",
              "      <th>F4_1+GGCTGTGAGGAACATT-1</th>\n",
              "      <th>F4_1+ACGTACACACCAGCGT-1</th>\n",
              "      <th>F4_1+ACGGTCGCAGATCACT-1</th>\n",
              "      <th>F4_1+CAACGATCAAATTGGA-1</th>\n",
              "      <th>F4_1+GTGGCGTAGCCTGCCA-1</th>\n",
              "      <th>F4_1+TAACCAGAGCTAGAAT-1</th>\n",
              "      <th>F4_1+TTACCATTCTGATGGT-1</th>\n",
              "      <th>F4_1+CAGCAATGTCTCGGAC-1</th>\n",
              "      <th>F4_1+AGACCCGTCGCTGATA-1</th>\n",
              "      <th>F4_1+ACCCAAAGTCGTGGAA-1</th>\n",
              "      <th>F4_1+GGGTTTAAGGTCGTCC-1</th>\n",
              "      <th>F4_1+CAACGATGTAGGAGTC-1</th>\n",
              "      <th>F4_1+AGGGAGTGTAGCGAGT-1</th>\n",
              "      <th>F4_1+TGCAGTACAAGACCGA-1</th>\n",
              "      <th>F4_1+GTCGCGAGTGGGTCAA-1</th>\n",
              "      <th>F4_1+GACCTTCCAGAGCCCT-1</th>\n",
              "      <th>F4_1+CGAGTTAGTCGTTATG-1</th>\n",
              "      <th>F4_1+CTACCTGCATCCGATA-1</th>\n",
              "      <th>F4_1+GGGTGTCTCCACGGGT-1</th>\n",
              "      <th>F4_1+GGCTGTGGTGCATTAC-1</th>\n",
              "      <th>F4_1+TCCCATGAGGGTCACA-1</th>\n",
              "      <th>...</th>\n",
              "      <th>F4_5+ATGCGATTCTCGGCTT-1</th>\n",
              "      <th>F4_5+ATTCCATAGCATGCAG-1</th>\n",
              "      <th>F4_5+CAAGCTATCTGGCCGA-1</th>\n",
              "      <th>F4_5+CCGTAGGAGTAAACGT-1</th>\n",
              "      <th>F4_5+CGGAGAAGTGTTCCAA-1</th>\n",
              "      <th>F4_5+CTCCCAAGTGGAGGTT-1</th>\n",
              "      <th>F4_5+CTGCGAGGTAAGTTAG-1</th>\n",
              "      <th>F4_5+CTTCAATTCGCGAAGA-1</th>\n",
              "      <th>F4_5+GATGAGGGTGTCTCCT-1</th>\n",
              "      <th>F4_5+GCAACCGAGTAGCAAT-1</th>\n",
              "      <th>F4_5+GGGTAGATCCTTCACG-1</th>\n",
              "      <th>F4_5+GTAGAAAAGACCATTC-1</th>\n",
              "      <th>F4_5+GTCAGCGGTATTGACC-1</th>\n",
              "      <th>F4_5+TAATTCCTCCCTTCCC-1</th>\n",
              "      <th>F4_5+TACAGGTCACAAATCC-1</th>\n",
              "      <th>F4_5+TCCTAATTCACACCCT-1</th>\n",
              "      <th>F4_5+TCGACCTGTGCATGTT-1</th>\n",
              "      <th>F4_5+TCGCAGGGTCGTACAT-1</th>\n",
              "      <th>F4_5+TCTCACGTCCGCAAAT-1</th>\n",
              "      <th>F4_5+TGCTCGTTCTTCTGGC-1</th>\n",
              "      <th>F4_5+TGGAACTGTCGAACAG-1</th>\n",
              "      <th>F4_5+TTACAGGTCCAGTACA-1</th>\n",
              "      <th>F4_5+TTCCGTGTCCTAAGTG-1</th>\n",
              "      <th>F4_5+ACGTAGTTCCTGTTGC-1</th>\n",
              "      <th>F4_5+ATCGCCTGTGATTAGA-1</th>\n",
              "      <th>F4_5+CAGTTAGAGAGGATGA-1</th>\n",
              "      <th>F4_5+CCACGTTAGGTTCATC-1</th>\n",
              "      <th>F4_5+CGTAAGTTCATAGAGA-1</th>\n",
              "      <th>F4_5+GAAATGAAGGGCCAAT-1</th>\n",
              "      <th>F4_5+GAACGTTTCTGTTGGA-1</th>\n",
              "      <th>F4_5+GAATCACCAATGCAAA-1</th>\n",
              "      <th>F4_5+GGAATGGGTTTCACTT-1</th>\n",
              "      <th>F4_5+GGAGATGCAAATGGTA-1</th>\n",
              "      <th>F4_5+GGGAGTAGTGCCGTAC-1</th>\n",
              "      <th>F4_5+GTCAAGTTCGCAGTCG-1</th>\n",
              "      <th>F4_5+GTCATTTTCACCCTGT-1</th>\n",
              "      <th>F4_5+TACGGTATCGCCGAGT-1</th>\n",
              "      <th>F4_5+TCCGGGAGTACGATCT-1</th>\n",
              "      <th>F4_5+TCGGATAAGCATGGGT-1</th>\n",
              "      <th>F4_5+TGAGTCATCACATACG-1</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>study</th>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>...</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>bendavid2021</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>sample_batch</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>...</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>sample</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>...</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "      <td>F4_5</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>cell_type</th>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Seam Cells</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Seam Cells</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>...</td>\n",
              "      <td>Seam Cells</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Excretory Gland</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Neuron</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Seam Cells</td>\n",
              "      <td>Pharyngeal Gland Cells</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Neuron</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Neuron</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Glia</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Neuron</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Neuron</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Hypodermis</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>cell_subtype</th>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Seam Cells</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Seam Cells</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Germline</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>...</td>\n",
              "      <td>Seam Cells</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Excretory Gland</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>ALM_PLM_AVM_PVM</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Seam Cells</td>\n",
              "      <td>Pharyngeal Gland Cells</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>AVG</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>AVJ</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Glia</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>AVF</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>AVK</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Hypodermis</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Hypodermis</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>sample_description</th>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>barcode</th>\n",
              "      <td>TGTAACGGTTAGCTAC-1</td>\n",
              "      <td>GGCAGTCCAGCCTATA-1</td>\n",
              "      <td>AAGTACCGTCATCCCT-1</td>\n",
              "      <td>AAGATAGTCCCTCTAG-1</td>\n",
              "      <td>ACCAAACCAGCTGTAT-1</td>\n",
              "      <td>CCGTTCATCACTCACC-1</td>\n",
              "      <td>ATGTCCCCAATCGCCG-1</td>\n",
              "      <td>GATTCGATCGTCAACA-1</td>\n",
              "      <td>TACGTCCAGTTGCTGT-1</td>\n",
              "      <td>GACCGTGTCCGTGCGA-1</td>\n",
              "      <td>CCGATGGTCTTACCAT-1</td>\n",
              "      <td>TCTGTCGTCTCAATCT-1</td>\n",
              "      <td>GACCGTGAGCGTGTTT-1</td>\n",
              "      <td>AGATGCTGTAGCTGCC-1</td>\n",
              "      <td>ACACCAACAGACGGAT-1</td>\n",
              "      <td>GCGGATCTCCCATGGG-1</td>\n",
              "      <td>TACCGAAAGCTGTGCC-1</td>\n",
              "      <td>AGCTTCCCAGATTAAG-1</td>\n",
              "      <td>CATTGCCAGGTACAAT-1</td>\n",
              "      <td>GGCTGTGAGGAACATT-1</td>\n",
              "      <td>ACGTACACACCAGCGT-1</td>\n",
              "      <td>ACGGTCGCAGATCACT-1</td>\n",
              "      <td>CAACGATCAAATTGGA-1</td>\n",
              "      <td>GTGGCGTAGCCTGCCA-1</td>\n",
              "      <td>TAACCAGAGCTAGAAT-1</td>\n",
              "      <td>TTACCATTCTGATGGT-1</td>\n",
              "      <td>CAGCAATGTCTCGGAC-1</td>\n",
              "      <td>AGACCCGTCGCTGATA-1</td>\n",
              "      <td>ACCCAAAGTCGTGGAA-1</td>\n",
              "      <td>GGGTTTAAGGTCGTCC-1</td>\n",
              "      <td>CAACGATGTAGGAGTC-1</td>\n",
              "      <td>AGGGAGTGTAGCGAGT-1</td>\n",
              "      <td>TGCAGTACAAGACCGA-1</td>\n",
              "      <td>GTCGCGAGTGGGTCAA-1</td>\n",
              "      <td>GACCTTCCAGAGCCCT-1</td>\n",
              "      <td>CGAGTTAGTCGTTATG-1</td>\n",
              "      <td>CTACCTGCATCCGATA-1</td>\n",
              "      <td>GGGTGTCTCCACGGGT-1</td>\n",
              "      <td>GGCTGTGGTGCATTAC-1</td>\n",
              "      <td>TCCCATGAGGGTCACA-1</td>\n",
              "      <td>...</td>\n",
              "      <td>ATGCGATTCTCGGCTT-1</td>\n",
              "      <td>ATTCCATAGCATGCAG-1</td>\n",
              "      <td>CAAGCTATCTGGCCGA-1</td>\n",
              "      <td>CCGTAGGAGTAAACGT-1</td>\n",
              "      <td>CGGAGAAGTGTTCCAA-1</td>\n",
              "      <td>CTCCCAAGTGGAGGTT-1</td>\n",
              "      <td>CTGCGAGGTAAGTTAG-1</td>\n",
              "      <td>CTTCAATTCGCGAAGA-1</td>\n",
              "      <td>GATGAGGGTGTCTCCT-1</td>\n",
              "      <td>GCAACCGAGTAGCAAT-1</td>\n",
              "      <td>GGGTAGATCCTTCACG-1</td>\n",
              "      <td>GTAGAAAAGACCATTC-1</td>\n",
              "      <td>GTCAGCGGTATTGACC-1</td>\n",
              "      <td>TAATTCCTCCCTTCCC-1</td>\n",
              "      <td>TACAGGTCACAAATCC-1</td>\n",
              "      <td>TCCTAATTCACACCCT-1</td>\n",
              "      <td>TCGACCTGTGCATGTT-1</td>\n",
              "      <td>TCGCAGGGTCGTACAT-1</td>\n",
              "      <td>TCTCACGTCCGCAAAT-1</td>\n",
              "      <td>TGCTCGTTCTTCTGGC-1</td>\n",
              "      <td>TGGAACTGTCGAACAG-1</td>\n",
              "      <td>TTACAGGTCCAGTACA-1</td>\n",
              "      <td>TTCCGTGTCCTAAGTG-1</td>\n",
              "      <td>ACGTAGTTCCTGTTGC-1</td>\n",
              "      <td>ATCGCCTGTGATTAGA-1</td>\n",
              "      <td>CAGTTAGAGAGGATGA-1</td>\n",
              "      <td>CCACGTTAGGTTCATC-1</td>\n",
              "      <td>CGTAAGTTCATAGAGA-1</td>\n",
              "      <td>GAAATGAAGGGCCAAT-1</td>\n",
              "      <td>GAACGTTTCTGTTGGA-1</td>\n",
              "      <td>GAATCACCAATGCAAA-1</td>\n",
              "      <td>GGAATGGGTTTCACTT-1</td>\n",
              "      <td>GGAGATGCAAATGGTA-1</td>\n",
              "      <td>GGGAGTAGTGCCGTAC-1</td>\n",
              "      <td>GTCAAGTTCGCAGTCG-1</td>\n",
              "      <td>GTCATTTTCACCCTGT-1</td>\n",
              "      <td>TACGGTATCGCCGAGT-1</td>\n",
              "      <td>TCCGGGAGTACGATCT-1</td>\n",
              "      <td>TCGGATAAGCATGGGT-1</td>\n",
              "      <td>TGAGTCATCACATACG-1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>7 rows × 55508 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                              F4_1+TGTAACGGTTAGCTAC-1  ...                            F4_5+TGAGTCATCACATACG-1\n",
              "study                                                    bendavid2021  ...                                       bendavid2021\n",
              "sample_batch                                                     F4_1  ...                                               F4_5\n",
              "sample                                                           F4_1  ...                                               F4_5\n",
              "cell_type                                                   Intestine  ...                                         Hypodermis\n",
              "cell_subtype                                                Intestine  ...                                         Hypodermis\n",
              "sample_description  C. elegans F4 segregants collected at the L2 l...  ...  C. elegans F4 segregants collected at the L2 l...\n",
              "barcode                                            TGTAACGGTTAGCTAC-1  ...                                 TGAGTCATCACATACG-1\n",
              "\n",
              "[7 rows x 55508 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "inEMlEhufC5r",
        "outputId": "9abe1272-0de9-4a9b-95ec-573971224cbf"
      },
      "source": [
        "### for this example we keep only genes with at least 10000 UMIs\n",
        "min_gene_counts=10000\n",
        "adata.var['gene_counts']=np.squeeze(np.asarray(adata.X.sum(0)))\n",
        "adata=adata[:,adata.var.gene_counts>min_gene_counts]\n",
        "adata"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "View of AnnData object with n_obs × n_vars = 55508 × 1094\n",
              "    obs: 'study', 'sample_batch', 'sample', 'cell_type', 'cell_subtype', 'sample_description', 'barcode'\n",
              "    var: 'gene_id', 'gene_name', 'wbps_transcript_id', 'chromosome_name', 'start_position', 'end_position', 'strand', 'external_gene_id', 'external_transcript_id', 'wormbase_locus', 'wormbase_gseq', 'gene_counts'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "cht4jflKfC5s",
        "outputId": "402b485e-7f0a-4f0e-b621-ce1ca48aae88"
      },
      "source": [
        "## register adata with SCVI, for more information see\n",
        "## https://docs.scvi-tools.org/en/stable/api/reference/scvi.data.setup_anndata.html\n",
        "adata.layers[\"counts\"] = adata.X.copy().tocsr() # converts to CSR format, preserve counts\n",
        "scvi.data.setup_anndata(adata, \n",
        "                        layer=\"counts\", \n",
        "                        batch_key='sample_batch')"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\u001b[34mINFO    \u001b[0m Using batches from adata.obs\u001b[1m[\u001b[0m\u001b[32m\"sample_batch\"\u001b[0m\u001b[1m]\u001b[0m                                        \n",
            "\u001b[34mINFO    \u001b[0m No label_key inputted, assuming all cells have same label                           \n",
            "\u001b[34mINFO    \u001b[0m Using data from adata.layers\u001b[1m[\u001b[0m\u001b[32m\"counts\"\u001b[0m\u001b[1m]\u001b[0m                                              \n",
            "\u001b[34mINFO    \u001b[0m Computing library size prior per batch                                              \n",
            "\u001b[34mINFO    \u001b[0m Successfully registered anndata object containing \u001b[1;36m55508\u001b[0m cells, \u001b[1;36m1094\u001b[0m vars, \u001b[1;36m5\u001b[0m batches,\n",
            "         \u001b[1;36m1\u001b[0m labels, and \u001b[1;36m0\u001b[0m proteins. Also registered \u001b[1;36m0\u001b[0m extra categorical covariates and \u001b[1;36m0\u001b[0m extra\n",
            "         continuous covariates.                                                              \n",
            "\u001b[34mINFO    \u001b[0m Please do not further modify adata until model is trained.                          \n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "oypEOE5ofC5s",
        "outputId": "258ca058-dbd7-4d02-feeb-9978803b7014"
      },
      "source": [
        "### DEFINE MODEL\n",
        "# these hyperparameters are fine for a small dataset, with a few batches \n",
        "# if integration is a problem then you can try increasing the layers to 3\n",
        "# and hieen units to 256\n",
        "\n",
        "model = scvi.model.SCVI(adata,\n",
        "                        n_hidden=128,\n",
        "                        n_layers=2,\n",
        "                        gene_likelihood='nb',\n",
        "                        dispersion='gene-batch'\n",
        "                        )\n",
        "\n",
        "# MODEL TRAINING\n",
        "# this model will train quickly even without a GPU, 25 epochs is not quite enough to\n",
        "# finish training, but this notebook is meant to run quickly just for showing the entire \n",
        "# data generation pipeline\n",
        "\n",
        "model.train(check_val_every_n_epoch =1,\n",
        "            # use_gpu=True,\n",
        "            max_epochs = 25,\n",
        "            plan_kwargs={'lr':1e-3})\n",
        "\n",
        "\n",
        "train_test_results = model.history['elbo_train']\n",
        "train_test_results['elbo_validation'] = model.history['elbo_validation']\n",
        "train_test_results.plot(logy=True)\n",
        "### MAKE SURE THE MODEL FINISHED TRAINING FOR BEST RESULTS\n",
        "display(train_test_results)\n"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "GPU available: True, used: True\n",
            "TPU available: False, using: 0 TPU cores\n",
            "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "Epoch 25/25: 100%|██████████| 25/25 [01:38<00:00,  3.96s/it, loss=573, v_num=1]\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>elbo_train</th>\n",
              "      <th>elbo_validation</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>epoch</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>791.531</td>\n",
              "      <td>713.661</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>698.824</td>\n",
              "      <td>694.957</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>683.969</td>\n",
              "      <td>683.079</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>673.916</td>\n",
              "      <td>673.483</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>664.885</td>\n",
              "      <td>669.278</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>657.165</td>\n",
              "      <td>655.469</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>650.134</td>\n",
              "      <td>647.986</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>642.678</td>\n",
              "      <td>643.525</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>637.132</td>\n",
              "      <td>637.071</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>632.294</td>\n",
              "      <td>633.142</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>627.501</td>\n",
              "      <td>628.212</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>623.777</td>\n",
              "      <td>625.189</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>620.613</td>\n",
              "      <td>622.67</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>617.397</td>\n",
              "      <td>619.738</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>615.288</td>\n",
              "      <td>618.452</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>613.058</td>\n",
              "      <td>615.89</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>610.729</td>\n",
              "      <td>615.113</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>608.791</td>\n",
              "      <td>612.513</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>607.394</td>\n",
              "      <td>611.738</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>606.511</td>\n",
              "      <td>611.053</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>605.116</td>\n",
              "      <td>610.376</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>604.467</td>\n",
              "      <td>609.345</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>603.859</td>\n",
              "      <td>608.432</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>603.124</td>\n",
              "      <td>608.836</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>601.925</td>\n",
              "      <td>607.169</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "      elbo_train elbo_validation\n",
              "epoch                           \n",
              "0        791.531         713.661\n",
              "1        698.824         694.957\n",
              "2        683.969         683.079\n",
              "3        673.916         673.483\n",
              "4        664.885         669.278\n",
              "5        657.165         655.469\n",
              "6        650.134         647.986\n",
              "7        642.678         643.525\n",
              "8        637.132         637.071\n",
              "9        632.294         633.142\n",
              "10       627.501         628.212\n",
              "11       623.777         625.189\n",
              "12       620.613          622.67\n",
              "13       617.397         619.738\n",
              "14       615.288         618.452\n",
              "15       613.058          615.89\n",
              "16       610.729         615.113\n",
              "17       608.791         612.513\n",
              "18       607.394         611.738\n",
              "19       606.511         611.053\n",
              "20       605.116         610.376\n",
              "21       604.467         609.345\n",
              "22       603.859         608.432\n",
              "23       603.124         608.836\n",
              "24       601.925         607.169"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAEKCAYAAADjDHn2AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3deXhURaL+8W91Z+2sJCGsWRAUEJRVUVwQcUMRRVFRR0HnquB2nXtn7sz8xjujo94ZFXVcUdxwF1cUBHVUQFFUFgGRXVkSthBCNpJO0kn9/jhNCDE7JJ2k38/z9JPuc06frkM/5E1Vnaoy1lpERETq4gp0AUREpPVTWIiISL0UFiIiUi+FhYiI1EthISIi9VJYiIhIvRQWIiJSL4WFiIjUKyTQBaiLMeZi4AIgFnjeWvtpgIskIhKUTENGcBtjfgf8B2CBH4HrrLXeRn+YMS8AY4Asa23/avvOAx4F3MBz1tp/VtnXAZhqrf1tbedOSkqy6enpjS2SiEhQW7ZsWba1tmN9x9UbFsaYbsAi4FhrbbEx5i1grrV2RpVjkoFia21BlW29rLWbqp3rdKAQeLlqWBhj3MAG4GwgE1gCXGmtXePf/xDwmrV2eW3lHDp0qF26dGl91ysiIlUYY5ZZa4fWd1xD+yxCgEhjTAjgAXZU2z8CmGWMCfd/+A3A49VPYq39Esip4fwnApustb9Ya0uBN4GLjON+YF5dQSEiIs2r3rCw1m4HpgLbgJ1AXvW+A2vt28AnwExjzNXA9cBljShHNyCjyutM/7bbgLOA8caYyTW90RhzoTFmel5eXiM+TkREGqPesPD3F1wE9AC6AlHGmN9UP85a+wDgBaYBY621hYdbOGvtY9baIdbaydbap2s5Zra19sa4uLjD/TgREalFQ+6GOgvYbK3dA2CMeQ8YDrxa9SBjzGlAf+B94G/ArY0ox3Ygpcrr7v5tItKKlJWVkZmZidfb6PtbJMAiIiLo3r07oaGhTXp/Q8JiG3CSMcYDFAOjgEN6ko0xg4DpOHc6bQZeM8bca629s4HlWAIcbYzpgRMSE4CrGvheEWkhmZmZxMTEkJ6ejjEm0MWRBrLWsnfvXjIzM+nRo0eTztGQPovvgHeA5Ti3zbpwgqEqD3C5tfZna20FcC2wtfq5jDFvAIuB3saYTGPMb/2f4cOpiXwCrAXestb+1KQrEpFm4/V6SUxMVFC0McYYEhMTD6tG2KBBedbav+E0LdW2/+tqr8uAZ2s47so6zjEXmNuQ8ohI4Cgo2qbD/d6CfrqPN7/fxjvLMgNdDBGRVq1VT/fREt7/YTsV1jJ+SPdAF0VEpNUK+ppFemIUm7OLAl0METkM6enpZGdns2XLFvr371//G+rxr3/9i6Kixv9e+Otf/8pnn3122J/fGgV9WKQlecguLKGwxBfooohIK1FXWJSXl9f6vr///e+cddZZzVWsgAr6Zqj0xCgAtu7dT7+uGtgn0lB3z/6JNTvyj+g5j+0ay98u7FfnMa+++iqPPfYYpaWlDBs2jKeeeuqQ/T6fj6uvvprly5fTr18/Xn75ZTweD59//jm///3v8fl8nHDCCUybNo3w8PBfnf+xxx5jx44djBw5kqSkJObPn090dDQ33XQTn332GU8++SRffPEFs2fPpri4mOHDh/PMM89gjGHSpEmMGTOG8ePHk56ezsSJE5k9ezZlZWW8/fbb9OnT54j+e7Uk1SwSPQBs3aumKJHWbu3atcycOZOvv/6aFStW4Ha7ee211w45Zv369dx8882sXbuW2NhYnnrqKbxeL5MmTWLmzJn8+OOP+Hw+pk2bVuNn3H777XTt2pX58+czf/58APbv38+wYcNYuXIlp556KrfeeitLlixh9erVFBcXM2fOnBrPlZSUxPLly5kyZQpTp049sv8YLSzoaxZp/prFlr37A1wSkbalvhpAc/j8889ZtmwZJ5xwAgDFxcUkJycfckxKSgqnnHIKAL/5zW947LHHOPvss+nRowfHHHMMABMnTuTJJ5/kjjvuaNDnut1uLr300srX8+fP54EHHqCoqIicnBz69evHhRde+Kv3XXLJJQAMGTKE9957r/EX3IoEfVhEh4eQFB3OVnVyi7R61lomTpzIP/7xj0O2z5gxo/J59fEER2JcSEREBG63G3AGJt58880sXbqUlJQU7rrrrloHux1o5nK73fh8bbtfNOiboQB6JHlUsxBpA0aNGsU777xDVlYWADk5OWzdeuhkEdu2bWPx4sUAvP7665x66qn07t2bLVu2sGmTs8TOK6+8wogRI2r9nJiYGAoKCmrcdyAYkpKSKCws5J133jns62oLFBY4TVHqsxBp/Y499ljuvfdezjnnHI4//njOPvtsdu7cecgxvXv35sknn6Rv377s27ePKVOmEBERwYsvvshll13Gcccdh8vlYvLkGlc9AODGG2/kvPPOY+TIkb/aFx8fzw033ED//v0599xzK5vE2rsGLavaFhzOSnlPfLGRqZ9uYO3fzyMyzH2ESybSfqxdu5a+ffsGuhjSRDV9f0d6pbx27UAn99YcNUWJiNQk6Du44eBYiy3ZRfTpHBvg0ohISxk3bhybN28+ZNv999/PueeeG6AStV4KCyC1cqyFahYiweT9998PdBHaDDVDAXGRoSREhbFFndwiIjVSWPilJXpUsxARqYXCwi9dt8+KiNRKYeGXnhjFjrxivGW1zygpIhKsFBZ+6UkerIXMfapdiLQ1R3o9i8Y644wzODDO6/zzzyc3N/dXx9x11131TiY4a9Ys1qxZU/m6Na2PobDwS6ty+6yISFPNnTuX+Pj4Jr23eli0pvUxdOusX7r/9lnNESXSQPP+BLt+PLLn7HwcjP5nnYc093oWH3/8Mc8//zxvv/02AAsWLGDq1KnMmTOHKVOmsGTJEoqLixk/fjx33333r96fnp7O0qVLSUpK4r777uOll14iOTmZlJQUhgwZAsCzzz7L9OnTKS0tpVevXrzyyiusWLGCDz/8kIULF3Lvvffy7rvvcs8991Suj1Fb+Vtq3QzVLPziPWHERYYqLERasZZYz+Kss87iu+++Y/9+53fBzJkzmTBhAgD33XcfS5cuZdWqVSxcuJBVq1bVWtZly5bx5ptvsmLFCubOncuSJUsq911yySUsWbKElStX0rdvX55//nmGDx/O2LFjefDBB1mxYgU9e/asPL6+8rfEuhmqWVSRnujRHVEiDVVPDaA5tMR6FiEhIZx33nnMnj2b8ePH89FHH/HAAw8A8NZbbzF9+nR8Ph87d+5kzZo1HH/88TWW9auvvmLcuHF4PE6rxdixYyv3rV69mjvvvJPc3FwKCwvrHTG+fv36OsvfEutmKCyqSEuM4oeMfYEuhojUoqXWs5gwYQJPPPEECQkJDB06lJiYGDZv3szUqVNZsmQJHTp0YNKkSbWuY1GfSZMmMWvWLAYMGMCMGTNYsGBBk85zQEusm6FmqCrSEz1s31dMqa8i0EURkRq01HoWI0aMYPny5Tz77LOVTVD5+flERUURFxfH7t27mTdvXp1lPf3005k1axbFxcUUFBQwe/bsyn0FBQV06dKFsrKyQ5rRaltHo7Hlbw4KiyrSEqOo0O2zIq1WS61n4Xa7GTNmDPPmzWPMmDEADBgwgEGDBtGnTx+uuuqqyqau2gwePJgrrriCAQMGMHr06EPWvbjnnnsYNmwYp5xyyiGd0RMmTODBBx9k0KBB/Pzzz5XbG1v+5qD1LKpYtnUfl077hhcnncDIPsn1v0EkyGg9i7ZN61kcIbp9VkSkZurgriIhKoyY8BDdESUSJLSeRcMpLKowxpCW5FHNQqQO1tom3WHUGgXTehaH2+WgZqhq0jT7rEitIiIi2Lt372H/4pGWZa1l7969RERENPkcqllUk57o4ZPVu/CVVxDiVpaKVNW9e3cyMzPZs2dPoIsijRQREUH37t2b/H6FRTVpiVH4Kizbc4srJxcUEUdoaCg9evQIdDEkAPSnczXpB2afVVOUiEglhUU1B26f1RKrIiIHKSyq6RgTjifMrXUtRESqUFhUY4zx3xGlmoWIyAEKixqkJ2qshYhIVQqLGqQlRpGRU0x5he4lFxEBhUWN0hM9lJZXsDOvONBFERFpFRQWNTgwvkKd3CIiDoVFDdKTNPusiEhVCosadIqJIDzEpTuiRET8FBY1cLkMaYkejeIWEfFTWNQiXWMtREQqKSxqkZ7kTFVeodtnRUQUFrVJS/RQ4qtgd4E30EUREQk4hUUt0nX7rIhIJYVFLdI0+6yISCWFRS26xEUS5nbpjigRERQWtXK7DCkJkWzJVs1CRERhUYf0xCiN4hYRQWFRJ2ddiyKs1e2zIhLcFBZ16JHkobisnD0FJYEuiohIQCks6lA5+6w6uUUkyCks6lA51kL9FiIS5BQWdegaH0GIy2ishYgEPYVFHULcLlISNPusiIjCoh5piR7VLEQk6Cks6pGeGMXWbN0+KyLBTWFRj7REDwUlPvbuLw10UUREAkZhUY8Dd0SpKUpEgpnCoh7pSZqqXEREYVGPbvGRuHX7rIgEOYVFPcJCXHSLj9TtsyIS1BQWDaDbZ0Uk2CksGsCZqlw1CxEJXgqLBkhL9JBXXEZukW6fFZHgpLBogHTNPisiQU5h0QDpSR4ALbEqIkFLYdEA3Tt4MEZTlYtI8FJYNEBEqJuucZFsVTOUiAQphUUDpSd5VLMQkaClsGigtMQo1SxEJGgpLAAaMP14eqKHnP2l5BWXtUCBRERal+AOi4oKmPM7+Oxv9R6a5r99dptqFyIShII7LFwusBXwzeOQubTOQw+OtVC/hYgEn+AOC4Cz74GYrjDrZijz1npYaoIz1kJzRIlIMFJYRMTC2Echez0svL/WwyLD3HSOjdAobhEJSgoLgF5nwcDfwNePwvbltR6WlujRKG4RCUoKiwPOvQ+ik+GDW8BXUuMhPZI0+6yIBCeFxQGR8XDho5C1Br6cWuMhaYlRZBeWUFjia+HCiYgElsKiqmPOheMnwFcPwc6Vv9qdnqhObhEJTgqL6s77B0QlwaxbwHfo+hUHxlpoJLeIBBuFRXWeBBjzCOz+ERY9fMiuNH/NQmMtRCTYKCxq0ucC6D8evnwQdq2u3BwVHkLHmHC2ZqtmISLBpVWHhTHmYmPMs8aYmcaYc1r0w0c/AJEd4IObofzgfFDpiZp9VkSCT71hYYzpbYxZUeWRb4y5oykfZox5wRiTZYxZXcO+84wx640xm4wxfwKw1s6y1t4ATAauaMpnNllUIlzg7+j++l+VmzX7rIgEo3rDwlq73lo70Fo7EBgCFAHvVz3GGJNsjImptq1XDaebAZxXfaMxxg08CYwGjgWuNMYcW+WQO/37W9axF8GxF8PCByBrLeDULHbleykuLW/x4oiIBEpjm6FGAT9ba7dW2z4CmGWMCQcwxtwAPF79zdbaL4GcGs57IrDJWvuLtbYUeBO4yDjuB+ZZa2sfWt2czp8K4THO3FHlPtKT/HdE5agpSkSCR2PDYgLwRvWN1tq3gU+AmcaYq4Hrgcsacd5uQEaV15n+bbcBZwHjjTGTa3qjMeZCY8z0vLy8RnxcI0R3hPMfhB3LYfHjlbPPbt6jsBCR4NHgsDDGhAFjgbdr2m+tfQDwAtOAsdbawsMtnLX2MWvtEGvtZGvt07UcM9tae2NcXNzhflzt+l0CfcbA/H/Qy2wnISqMqZ+uJ9+rhZBEJDg0pmYxGlhurd1d005jzGlAf5z+jPpXEzrUdiClyuvu/m2tgzFwwcMQ5iFi7u08deUAtu4t4o43V1BeUf8qeyIibV1jwuJKamiCAjDGDAKmAxcB1wGJxph7G3HuJcDRxpge/hrMBODDRry/+cV0cm6nzVzCSVkz+dvYfnyxLosHP1kf6JKJiDS7BoWFMSYKOBt4r5ZDPMDl1tqfrbUVwLVA9U5wjDFvAIuB3saYTGPMbwGstT7gVpx+j7XAW9banxp7Mc3uuMvgmNHwxb1cc1QRVw9L5emFP/PBitZTCRIRaQ7G2vbRjDJ06FC7dGndS6MeEfk74ZnTwLgovXoWv/kwj5UZubw9+WSO7x7f/J8vInIEGWOWWWuH1ndcqx7B3SrFdoGJc8Bawl65kOnnRJIUHc6NLy8jK7/2ZVlFRNoyhUVTJPeB6+aCO5T4t8bx8gUe8orLmPzqMkp8GqwnIu2PwqKpko6GSR9BaCQ9507gubNDWL4tl7+8v5r20rQnInKAwuJwJPZ0AiMsmlO+uZ77TizlnWWZvPD1lkCXTETkiFJYHK6EHk5gRMRz1frbmdxzL/d9tIYvN+wJdMlERI4YhcWR0CENrpuL8STyxz1/5uLEDG59fTmbszUliIi0DwqLIyWuuxMYMZ2ZWnI3J5o13PDyUgo0JYiItAMKiyMptitM+ghXXArPuP5J573f8Z+aEkRE2gGFxZEW0xkmzcGdeBQzwqdStuEzpn6qKUFEpG1TWDSH6GSYOBt3x168EP4wa798V1OCiEibprBoLlFJmIlzcHfqw7NhjzD33Rf5aNXOQJdKRKRJFBbNyZOAa+KHmM79edL9MItn3s/jn23QoD0RaXMUFs0tsgMhkz7A9BrFvaEvkrTgf/jDm0s0LYiItCkKi5YQEYf7qpnY037PlSHzuWrtFG55ei7ZhSWBLpmISIMoLFqKy4UZ9b9w2UscH7qD/9tzG395/AU27C4IdMlEROqlsGhp/S4m5MbPiIuN4fGSO3n1qXtZsD4r0KUSEamTwiIQOvUjfMpCKlJP4e/maTJevZmXv9oQ6FKJiNRKYREongQiJr5H6bBbucb9b3r/+xr++fZX+MorAl0yEZFfUVgEkjuEsNH3UTHuWQa7N3Pt6mu5e/pr5BVrPikRaV0UFq2Aa8DlhN7wKXGecP6y63c89eh9bNtbFOhiiYhUUli0Fl0HEnXLV5R0HsyfvY+w8PEb+G7T7kCXSkQEUFi0LtEdibvxI/KOv55rmEPcy6P49rW7sfmaJkREAkth0dq4Q4m75BGKxj5DZGQkJ218GPvwsfhmjIUVr4M3P9AlFJEgZNrLPEVDhw61S5cuDXQxjihrLW9//AVZ37zCJSHf0NXuhpAI6D0ajrscep0FIWGBLqaItGHGmGXW2qH1HRfSEoWRpjHGcPnoUXzfZxAXv7aMo0rWcG/6Wnpt/hR+eh8iO0C/cU5wpAwDlyqKItI8VLNoI/YUlHD7Gz+w+Je9XDW0C3cdu5uwNe/Auo/AVwzxqXDcZXDyreBJCHRxRaSNaGjNQn+KthEdY8J55bcncsvInry+dCfjPotm28jH4Q8bYdwzkHg0LHoEXr8cyryBLq6ItDMKizYkxO3iD+f24fmJQ8nIKeKCx7/i3z8XwYAJcM17cNlLkLkEPrgF2kmNUURaB4VFGzSqbyc+uv000hI93PDyUv4xb60zTcixY2HUX2H1O7DwgUAXU0TaEYVFG5WS4OGdycO5elgqzyz8hauf+46sAi+c+l8w4EpY8H+w+t1AF1NE2gmFRRsWEermvnHH8fDlA1iZmcsFjy3iy43ZcOGjkHoyzLoZMttvp7+ItByFRTtwyeDufHDLqcRGhHDtC9/z5w/XU3jxDIjpDG9cCbkZgS6iiLRxCot2onfnGD66/TRuGnEUM5dkcM4zP7Fk+DTweeGNCVCiFflEpOkUFu1IRKibP4/uy7tThhMZ5uayd/fxfOe/YrPWwrs3QEV5oIsoIm2UwqIdGpTagY9uP43JI3py34auTHVfDxvmwb//GuiiiUgbpbBopyJC3fxpdB/eu/kUPvFcyAzfObD4CYq+fSHQRRORNkhh0c4NTIlnzm2nknXK31hYcTxhH/+eH778MNDFEpE2RmERBCJC3fzP6P4kTHyNTFdXenw+mftfnaPlW0WkwRQWQeS4nql0mfIhYaGhXL7xvxn/8By+WKfV+ESkfgqLIBPe8Sg8184kzZ3DgxUPceOMb7nl9eXsytPkgyJSO4VFMEo9CddFjzOw/Ec+SHuXz9bs5KyHF/L8os3OHFMiItUoLILVgAlw2u/pt/sDVqY+yvld93PPnDWMfeJrlm/bF+jSiUgro7AIZmfeCRdPIyJnHffvmcLHQ5eTV1jMpdO+4c/v/UhuUWmgSygirYTCIpgZAwOvglu+x/Q6iz6rp/Jl4n38aVA5by3N4MyHFvL20gzay2qKItJ0CgtxJhy84lW4bAbu/ExuWnc93578PUcnhvGHd1ZxxTPfsmG35pYSCWYKC3EYA/3GwS3fQ79xdFz2CG/yJ54dZdiYVcD5j37FP+atpajUF+iSikgAKCzkUFGJcOmzcOVMTHEuZ39zNd8MWcgVAzvyzMJfOOuhhfx7jcZmiAQbhYXUrPd5cMu3MOgaIpc8wX27buLjcW5iI0O54eWl3PbGD+wtLAl0KUWkhSgspHYRcTD2Mbj2A6jw0WfeFXzU6wP+NLIbH6/eydmPfMnslTvUAS4SBBQWUr+jzoCbF8OwKbiXPs/kFeP4fsgXDI/Zw21v/MBNrywjK18jwEXaM9Ne/iocOnSoXbpU6003u4wlsPgJWPcRVJSxK24A/8o5mfnu4fzhwiFcOrgbxphAl1JEGsgYs8xaO7Te4xQW0iSFe2DlG7D8Zdi7kSLj4b2yk9nQ/VJumnAJ3eIjA11CEWkAhYW0DGth22LsshmUr55FSEUJP9ke7O93FUPH3IjLEx/oEopIHRQW0vKKc9n37Wvkf/McaWW/UGLCKet9EdHDfwspw5yxHCLSqjQ0LNTBLUdOZDwdRt5C6p+X8cnwN/mw4jTM2g/hhXOxL46GHSsCXUIRaSKFhRxxxuXi3HNGc+p/vcp/p77FnWXXkZuxBjv9DCpm3QIFGtQn0tYoLKTZdImLZNr1pzN8wh+5LuZppvvOp3zFm5Q9OojyLx8Gnwb1ibQVCgtpVsYYzj+uC+/dMZrUKx7ippgnWVDSB/cXd1P48BDKf/rQ6SQXkVZNYSEtwuUyjD6uC8/97goqJrzOX2LuYXuhxf32Nex+4hzKdqwKdBFFpA4KC2lRLpfh3H6dufe/biPz8k+ZFjWF0Ow1uKaPYOPz/0FpXlagiygiNdCtsxJQ1loWrd5E7tx7OK9oDl4TzvreN3PcJb8nPFwD+0Sam8ZZSJtirWXp0u9w//svDC5dSgadyehxGf3Ovo64rj0DXTyRdkthIW2StZa1X75LyNcPcUzpGgA2e47HM2QCnU6a4Ky3ISJHjMJC2rzNG1ez4fOXOGrnXI42mfhwk9vlVBJOuhpXnwsgPDrQRRRp8xQW0m7sKyzh0wVfUPrDm5zp+4puZi9lrgjocz6hA66AnmdCSFigiynSJikspN0pK6/gk9U7+HbBXPrs+Zgx7u+IN4WUh3fA3f9iGHAlpJyoOahEGkFhIe3aqsxcXl60kbzVnzDGfM15IcsJt15styGYk2+BvheBOyTQxRRp9RQWEhSy8r28+t023v92HSO8XzAl/GO6VeykIrY7rmE3weBrIVLTpIvURmEhQcVbVs6cVTuZsehnOu9eyOSweQxlDRWhUbgGXwsnTYYO6YEupkiro7CQoGStZcmWfbz49WYy1yzmevc8xroX4zIW+ozBnHyr+jVEqlBYSNDL3FfEK4u38vn3K7jEN5drQz8n2u6notsQXCffCn3Hql9Dgp7CQsSvqNTH+z9s541FaxmUM48bQj8mlV2Ux3THPeAy6DnKWclPt99KEFJYiFRjrWXRpmxmLPoZ18ZPmBTyKSe51uKmHBsahelxmjNmo+eZkNhLTVUSFBQWInXYnL2fVxZv5YuVm+hVtIKRIas5O3w1yWXbnQPiUqHnSCc4jhoBkR0CW2CRZqKwEGmA8grL95tzmLNqBx+v3oWnKINRoT8xLnY9x5b8QGhZIRgXdBtysNbRdRCEhAe66CJHhMJCpJF85RV8+4s/OH7aRUGRl5PDt3BNx02cZFcSm7MKYyvAHQ7dBjv9HAcemuBQ2iiFhchhKCuv4OtN2cxZtZNPftpFgddH94gSbkjdzoiIX0gpXIV710qoKHPekHg0pA6DlJOc8Eg6Wn0e0iYoLESOkBJfOYs2OsHx2ZrdFJT4CHO7OCU9iks6ZTE8bBMJOT9gMr6H4hznTZEJTmikDoPU4U5NxB0a2AsRqYHCQqQZlPoqWLZ1H/PXZzF/XRYbswoBSE3wMPKYJM7vVshg1hO6fQlkfAd7NzpvDIuBHqfBUSPhqDNU85BWQ2Eh0gIycopYsGEPC9Zl8fXP2XjLKggPcTG8ZyIj+yQzKsVFt/wf4Of58Mt82LfFeWNsdyc0eo6EHiMgumMAr0KCmcJCpIV5y8r5bnMO89dlMX99Flv3FgHQs2MUo/p24sw+yQyJzSN0ywL4ZQH8shC8uc6bOx0HPc9wah6pJ0OYJ1CXIUFGYSESYJuz9/PFOqe56rvNeykrt8RGhHD6MR0Z1TeZM3ol0iF/jb/WsQC2fet0mLvDnWaquO4Ql+L/6X8enwLRncDlDvTlSTuhsBBpRQpLfCzauIfP1zq1juzCUlwGBqd24My+yZzZJ5neHVyYbd/C5gWQvQnyMpyHN+/Qk7lCILarP0j8YRKf6owFST4WXK6AXKO0TQoLkVaqosLy4/Y8Pl+XxRfrdrN6ez4A3eIjObOPExwn9EggOtw/yaE3H/K3Q64/PPIyqzwyIH8H2HLn2Ig4/11YJzuPboM1gFDqpLAQaSN253uZvy6Lz9dlsWhjNsVl5RgDPZKi6Nc1jv5dY+nfLY5+XWOJ99Qw2WG5D3K3QuYS2PqN05yVvd7Zd2AAYerJkDYcup+gxaDkEAoLkTbIW1bO95tzWJGRy+rtefy0I5/tucWV+7t3iKR/1zj6d4ulnz9AkmMifn2i/dnOrbsHwmPnCqjwAQY69YfUk6DrQPAkOmNCIjscfGja9qCisBBpJ/btL+WnHfms3pFXGSCbs/dX7k+OCad/tziO6xbHgJQ4ju8eT1J0taan0v2wfRlsXQzbFkPG91C2nxqFxzq1jwMh4qkSJkm9Ie1kp59E2gWFhUg7VuAtY+3OAlZvz6sMkY1ZhRz479wtPpLjuzvBMaB7HP27xxEbUWUEebnP6e8ozrz5hKEAAAy1SURBVIHifVCcC0UHnu87uL3qNm8u2Arn/fGpkHaK07SVOhwSe2qQYRulsBAJMvtLfKzenseqzDxWZuayKjOPbTlFlfuPSoo6GCApcfTrGkdEaCNuwS33QdZPTtPWgUdRtrMvKtkJjrRTnJpHcj/dldVGKCxEhH37S/lxex6rMnNZmen83J1fAoDbZejbJYaBKfEMSunAwNR4eiRG4XI1sIZgLWRvhK1fO01bW76G/ExnX0Sc06meMgxiux3alBXZASLiFSathMJCRGq0O9/LyoxcVmbmsiIjl5UZeRSW+ACIiwxlQEo8g1LiGZgaz8Du8XSIasRys7nbDq15HJgb61dMLf0iCc72iHgncCL9PyPiDm4Li1KT1xGksBCRBimvsGzKKmRFxj5+2OYEyPrdBZX9Hz2SoirDo3+3OPp0jsET1sA7popznTuzauwHqaFfpHgflOTXfU5XSJUAqRIi1R+RHWreHhp5eP9g7YzCQkSarLDExyp/zeOHbc4ju9BpvjIG0hI89OkcS98usfTpEkPfzrF07xDZ8CasupT7nMDw5jph483zP6o8r227Nw983rrP7w53QiM62f/o7PyM6exMpRLdyf88GcJjDv96WrmGhoVuqBaRX4kOD2F4zySG90wCwFrL9txi1uzIZ92uAtbudH5+smZXZQ0kOjyE3p1j6NM5hj5dYjm2SwzHdIohJqKR63i4Q5xmKU9C0wpf5vWHTV0hkwuFe6BwF+zZAIW7Dy5kVVVoFMT4AyQ6GaI6+h9JVZ539I9X6dD05jFrnZDzlThB1gqb2VSzEJEmKyr1sWF3oRMeO/NZu6uAdTvzyff6Ko9Jig4nJSGS1AQPKR08pCZ46O5/3SUuEveRqI0cLmudJrCCXU6AFGb5n+92HgX+n0X+JrWauELAk3RomLhDoazY/yhyAqGsyP+6ynPfwYGXhEY5E0kmHeN/+J8nHAWhNQzAPExqhhKRgLDWsiPPyzp/7SMjp4iMfUVsyyliR66X8oqDv3NCXIau8f4gSYgkJcFDt/hIEqPCSYgKIzE6jISoMELdrejOqfIyKNoL+/f4H9n+R9XX/ucV5U4fSWgEhHogxP8zNPLgo+o2d6gz51f2BudOs7yMg59rXBCfdmiAHHgcxhrwCgsRaXV85RXszPOSkeOEhxMixU6g5BSxd39pje+LjQghMdofIFVCJCEqnMSoMDrHRZCS4KFzbETrqKkcKaX7Ye8mJziyNxwMkb2bDu2buWO1M319E6jPQkRanRC3i5QEDykJHobXsH9/iY+decXsLSwlZ38p2ftLySksJWd/CXv3O9u27i1i+bZc9hWVHlJLAQh1G7rFR1Z+Rqr/caD5K87TxtZBD4uCLgOcR1UVFU6tI3ujc3tybLdmL4rCQkRajajwEHolx9Aruf5jKyos+d4ysgtL2ZlXTEZOcWVtJSOniB9/3Elu0aGd1jERIU6fSYdIEqLCiI0IJSYihNjI0FqfR4W5Ma2tw9nlgg5pzuPos1rkIxUWItImuVyGeE8Y8Z4weiVH13hMvrfM38RVfEjfyc979rNsay4F3jJKfBV1f46BGH94HPgZGxFCdPjB19H+fdW3x0aGEh8Ziqc1Bk4jKSxEpN2KjQilX1dnHqzaeMvKKfD6KPCWke/1kV9cRoHXR7637FfPC0t85Ht97Mj1UlBSRqHXR4HXh6+i7r7fULchLjKUuMhQJ+AiQ4nz+F9HhhHvCSXe/7pTbARd4iKIiwxtVQGjsBCRoBYR6iYi1E3HmKatKGitxVtWQUGJEywHAsQJnzJyi8rIKy4jt7iMvKIycotL2V3gZf3uAvKKyigo8dV4Xk+Ym85xEXSNi6RLXARd4iPpWu1n5WqKLaBVh4Ux5mLgAiAWeN5a+2mAiyQicghjDJFhbiLD3CQ3YcB3WXkF+cVOoOwrKmVXXgk784rZket1fuZ52bBhD3sKS6h+82pMRAhd4yJ55bcnkhx75MdgVNWgsDDGxAPPAf0BC1xvrV3c2A8zxrwAjAGyrLX9q+07D3gUcAPPWWv/aa2dBcwyxnQApgIKCxFpV0LdLhKjw0msvmBVNaW+Cnbne9mZ5z00THK9xEY2/11eDa1ZPAp8bK0db4wJAzxVdxpjkoFia21BlW29rLWbqp1nBvAE8HK197uBJ4GzgUxgiTHmQ2vtGv8hd/r3i4gEpbCQg7cdB0K9wyKNMXHA6cDzANbaUmttbrXDRuDUAML977kBeLz6uay1XwI5NXzMicAma+0v1tpS4E3gIuO4H5hnrV1eS/kuNMZMz8vLq+9SRESkiRoyhr4HsAd40RjzgzHmOWNMVNUDrLVvA58AM40xVwPXA5c1ohzdgCrj2sn0b7sNOAsYb4yZXNMbrbWzrbU3xsXVfreDiIgcnoaERQgwGJhmrR0E7Af+VP0ga+0DgBeYBoy11hYebuGstY9Za4dYaydba58+3POJiEjTNCQsMoFMa+13/tfv4ITHIYwxp+F0gL8P/K2R5dgOVJ3YpLt/m4iItAL1hoW1dheQYYzp7d80ClhT9RhjzCBgOnARcB2QaIy5txHlWAIcbYzp4e9AnwB82Ij3i4hIM2rovL+3Aa8ZY1YBA4H/q7bfA1xurf3ZWlsBXAtsrX4SY8wbwGKgtzEm0xjzWwBrrQ+4FaffYy3wlrX2p6ZckIiIHHmaolxEJIg1dIryVrSiiIiItFbtpmZhjNlDDU1fDZQEZB/B4rQlwXztENzXH8zXDsF9/VWvPc1a27G+N7SbsDgcxpilDamGtUfBfO0Q3NcfzNcOwX39Tbl2NUOJiEi9FBYiIlIvhYVjeqALEEDBfO0Q3NcfzNcOwX39jb529VmIiEi9VLMQEZF6BXVYGGPOM8asN8ZsMsb8anLE9s4Ys8UY86MxZoUxpt2PaDTGvGCMyTLGrK6yLcEY829jzEb/zw6BLGNzqeXa7zLGbPd//yuMMecHsozNxRiTYoyZb4xZY4z5yRjzn/7t7f67r+PaG/3dB20zlH/BpQ1UWXAJuLLKgkvtnjFmCzDUWhsU95obY04HCoGXD6zUaIx5AMix1v7T/wdDB2vtHwNZzuZQy7XfBRRaa6cGsmzNzRjTBehirV1ujIkBlgEXA5No5999Hdd+OY387oO5ZlHjgksBLpM0o1oW37oIeMn//CWc/0jtTh0Lj7V71tqdBxZP86/muRZnvZx2/93Xce2NFsxhUduCS8HEAp8aY5YZY24MdGECpJO1dqf/+S6gUyALEwC3GmNW+Zup2l0zTHXGmHRgEPAdQfbdV7t2aOR3H8xhIXCqtXYwMBq4xd9UEbSs0yYbTO2y04CeODNJ7wQeCmxxmpcxJhp4F7jDWptfdV97/+5ruPZGf/fBHBZBv+CStXa7/2cWzqJVJwa2RAGx29+ue6B9NyvA5Wkx1trd1tpy/7ICz9KOv39jTCjOL8vXrLXv+TcHxXdf07U35bsP5rAI6gWXjDFR/g4v/GuqnwOsrvtd7dKHwET/84nABwEsS4s68IvSbxzt9Ps3xhjgeWCttfbhKrva/Xdf27U35bsP2ruhAPy3i/0LcAMvWGvvC3CRWowx5iic2gQ466y/3t6v37/41hk4M27uxln+dxbwFpCKM2vx5dbadtcRXMu1n4HTDGGBLcBNVdrw2w1jzKnAV8CPQIV/8//Dabtv1999Hdd+JY387oM6LEREpGGCuRlKREQaSGEhIiL1UliIiEi9FBYiIlIvhYWIiNRLYSHSChhjzjDGzAl0OURqo7AQEZF6KSxEGsEY8xtjzPf+NQCeMca4jTGFxphH/OsFfG6M6eg/dqAx5lv/ZG3vH5iszRjTyxjzmTFmpTFmuTGmp//00caYd4wx64wxr/lH34q0CgoLkQYyxvQFrgBOsdYOBMqBq4EoYKm1th+wEGd0NMDLwB+ttcfjjKA9sP014Elr7QBgOM5EbuDMCHoHcCxwFHBKs1+USAOFBLoAIm3IKGAIsMT/R38kzuRzFcBM/zGvAu8ZY+KAeGvtQv/2l4C3/fNxdbPWvg9grfUC+M/3vbU20/96BZAOLGr+yxKpn8JCpOEM8JK19s+HbDTmf6sd19Q5dEqqPC9H/z+lFVEzlEjDfQ6MN8YkQ+Uazmk4/4/G+4+5Clhkrc0D9hljTvNvvwZY6F+tLNMYc7H/HOHGGE+LXoVIE+gvF5EGstauMcbcibO6oAsoA24B9gMn+vdl4fRrgDPt9dP+MPgFuM6//RrgGWPM3/3nuKwFL0OkSTTrrMhhMsYUWmujA10OkeakZigREamXahYiIlIv1SxERKReCgsREamXwkJEROqlsBARkXopLEREpF4KCxERqdf/B4xnf3kVKNcgAAAAAElFTkSuQmCC\n",
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ]
          },
          "metadata": {
            "tags": [],
            "needs_background": "light"
          }
        }
      ]
    },
    {
      "cell_type": "raw",
      "metadata": {
        "id": "nIfOTJONfC5t"
      },
      "source": [
        "\"\"\"\n",
        "FOR HEATMAP: MEAN EXPRESSION PER CELL TYPE\n",
        "\n",
        "FOR RIDGEPLOT: BINNED EXPRESSION VALUES PER CELL TYPE\n",
        "Now that the scVI model is trained, we get the normalized expression for all cells,\n",
        "then use that to calculate binned expression values logspaced between 10^-9 and 1\n",
        "\n",
        "FOR SWARMPLOT:\n",
        "log2 fold change pairwise for each cell type\n",
        "\n",
        "\"\"\""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OQFXXSR0Dxx5"
      },
      "source": [
        "# Computes global DE (each cell type vs rest) \n",
        "\n",
        "**Make the gene expression heatmap anndata**\n",
        "This data is a 2D matrix of shape:\n",
        "$ x_{celltypes} \\times y_{genes} = x_{obs} \\times y_{var}  $ \n",
        "\n",
        "```\n",
        "anndata.obs = cell_types\n",
        "anndata.var = gene_id\n",
        "anndata.X = log10 scvi expression frequency values in the X field\n",
        "```"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 525
        },
        "id": "kIM9Ha2bmspT",
        "outputId": "e32a78f1-e5ce-4fd0-a41b-7145c11ad0dd"
      },
      "source": [
        "# perform DE on each cell type vs the rest of cells, this computes the expresssion (scale1)\n",
        "# in each celltype, used for the heatmap anndata, plus scale1, the p-values and lfc_median\n",
        "# for each cell type which are used for ranking the swarmplot\n",
        "\n",
        "de_global=model.differential_expression(\n",
        "    groupby='cell_type',\n",
        "    all_stats=False\n",
        ")\n",
        "\n",
        "# scvi currently puts the groups in a column named \"comparison\", eg\n",
        "# an entry would be \"Neurons vs Instestine\" but we need to split that into\n",
        "# one column for group1 and group2. Submitted a PR to change that:\n",
        "# https://github.com/YosefLab/scvi-tools/pull/1074\n",
        "de_global['group1']=de_global['comparison'].str.split(' vs ', expand=True)[0]\n",
        "de_global['group2']=de_global['comparison'].str.split(' vs ', expand=True)[1]\n",
        "de_global"
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "DE...: 100%|██████████| 19/19 [00:47<00:00,  2.50s/it]\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>proba_de</th>\n",
              "      <th>proba_not_de</th>\n",
              "      <th>bayes_factor</th>\n",
              "      <th>scale1</th>\n",
              "      <th>scale2</th>\n",
              "      <th>pseudocounts</th>\n",
              "      <th>delta</th>\n",
              "      <th>lfc_mean</th>\n",
              "      <th>lfc_median</th>\n",
              "      <th>lfc_std</th>\n",
              "      <th>lfc_min</th>\n",
              "      <th>lfc_max</th>\n",
              "      <th>is_de_fdr_0.05</th>\n",
              "      <th>comparison</th>\n",
              "      <th>group1</th>\n",
              "      <th>group2</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>WBGene00006789</th>\n",
              "      <td>0.9928</td>\n",
              "      <td>0.0072</td>\n",
              "      <td>4.926447</td>\n",
              "      <td>0.006613</td>\n",
              "      <td>0.000303</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>6.436170</td>\n",
              "      <td>5.791610</td>\n",
              "      <td>3.895258</td>\n",
              "      <td>-5.158882</td>\n",
              "      <td>22.906197</td>\n",
              "      <td>True</td>\n",
              "      <td>Body Wall Muscle vs Rest</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00006587</th>\n",
              "      <td>0.9924</td>\n",
              "      <td>0.0076</td>\n",
              "      <td>4.871977</td>\n",
              "      <td>0.004109</td>\n",
              "      <td>0.000293</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>6.182913</td>\n",
              "      <td>5.243339</td>\n",
              "      <td>4.344330</td>\n",
              "      <td>-5.738801</td>\n",
              "      <td>25.059113</td>\n",
              "      <td>True</td>\n",
              "      <td>Body Wall Muscle vs Rest</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00006754</th>\n",
              "      <td>0.9924</td>\n",
              "      <td>0.0076</td>\n",
              "      <td>4.871977</td>\n",
              "      <td>0.004897</td>\n",
              "      <td>0.000265</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>5.933573</td>\n",
              "      <td>5.165828</td>\n",
              "      <td>3.934400</td>\n",
              "      <td>-3.698731</td>\n",
              "      <td>24.183247</td>\n",
              "      <td>True</td>\n",
              "      <td>Body Wall Muscle vs Rest</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00000063</th>\n",
              "      <td>0.9922</td>\n",
              "      <td>0.0078</td>\n",
              "      <td>4.845800</td>\n",
              "      <td>0.015974</td>\n",
              "      <td>0.001214</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>5.911234</td>\n",
              "      <td>5.327559</td>\n",
              "      <td>3.687594</td>\n",
              "      <td>-6.126419</td>\n",
              "      <td>20.075211</td>\n",
              "      <td>True</td>\n",
              "      <td>Body Wall Muscle vs Rest</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00000779</th>\n",
              "      <td>0.9922</td>\n",
              "      <td>0.0078</td>\n",
              "      <td>4.845800</td>\n",
              "      <td>0.005760</td>\n",
              "      <td>0.000303</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>6.305322</td>\n",
              "      <td>5.683682</td>\n",
              "      <td>3.867903</td>\n",
              "      <td>-6.036078</td>\n",
              "      <td>24.724632</td>\n",
              "      <td>True</td>\n",
              "      <td>Body Wall Muscle vs Rest</td>\n",
              "      <td>Body Wall Muscle</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00004412</th>\n",
              "      <td>0.7376</td>\n",
              "      <td>0.2624</td>\n",
              "      <td>1.033532</td>\n",
              "      <td>0.007387</td>\n",
              "      <td>0.006496</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.441539</td>\n",
              "      <td>0.160584</td>\n",
              "      <td>1.335513</td>\n",
              "      <td>-3.704840</td>\n",
              "      <td>11.523404</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs Rest</td>\n",
              "      <td>XXX</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00006725</th>\n",
              "      <td>0.7358</td>\n",
              "      <td>0.2642</td>\n",
              "      <td>1.024252</td>\n",
              "      <td>0.005822</td>\n",
              "      <td>0.004723</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.554490</td>\n",
              "      <td>0.281458</td>\n",
              "      <td>1.313499</td>\n",
              "      <td>-3.148258</td>\n",
              "      <td>12.393562</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs Rest</td>\n",
              "      <td>XXX</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00004473</th>\n",
              "      <td>0.7338</td>\n",
              "      <td>0.2662</td>\n",
              "      <td>1.013989</td>\n",
              "      <td>0.006246</td>\n",
              "      <td>0.005168</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.453999</td>\n",
              "      <td>0.274692</td>\n",
              "      <td>1.121817</td>\n",
              "      <td>-2.542458</td>\n",
              "      <td>10.420438</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs Rest</td>\n",
              "      <td>XXX</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00004435</th>\n",
              "      <td>0.7148</td>\n",
              "      <td>0.2852</td>\n",
              "      <td>0.918812</td>\n",
              "      <td>0.005118</td>\n",
              "      <td>0.004067</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.596725</td>\n",
              "      <td>0.297470</td>\n",
              "      <td>1.287573</td>\n",
              "      <td>-2.833117</td>\n",
              "      <td>11.134089</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs Rest</td>\n",
              "      <td>XXX</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00004438</th>\n",
              "      <td>0.7148</td>\n",
              "      <td>0.2852</td>\n",
              "      <td>0.918812</td>\n",
              "      <td>0.001167</td>\n",
              "      <td>0.000934</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.629274</td>\n",
              "      <td>0.320423</td>\n",
              "      <td>1.471241</td>\n",
              "      <td>-1.935485</td>\n",
              "      <td>9.972778</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs Rest</td>\n",
              "      <td>XXX</td>\n",
              "      <td>Rest</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>20786 rows × 16 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                proba_de  proba_not_de  ...            group1  group2\n",
              "WBGene00006789    0.9928        0.0072  ...  Body Wall Muscle    Rest\n",
              "WBGene00006587    0.9924        0.0076  ...  Body Wall Muscle    Rest\n",
              "WBGene00006754    0.9924        0.0076  ...  Body Wall Muscle    Rest\n",
              "WBGene00000063    0.9922        0.0078  ...  Body Wall Muscle    Rest\n",
              "WBGene00000779    0.9922        0.0078  ...  Body Wall Muscle    Rest\n",
              "...                  ...           ...  ...               ...     ...\n",
              "WBGene00004412    0.7376        0.2624  ...               XXX    Rest\n",
              "WBGene00006725    0.7358        0.2642  ...               XXX    Rest\n",
              "WBGene00004473    0.7338        0.2662  ...               XXX    Rest\n",
              "WBGene00004435    0.7148        0.2852  ...               XXX    Rest\n",
              "WBGene00004438    0.7148        0.2852  ...               XXX    Rest\n",
              "\n",
              "[20786 rows x 16 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TDVNN86b-IIH",
        "outputId": "383f8b22-ee27-4eac-f898-759b4edb9452"
      },
      "source": [
        "# pivot the DE result dataframe to create a dataframe for the heatmap \n",
        "# with gene ids in the index and cell type name in the columns and \n",
        "# scale1 in the entries, then take the log10 of scale1\n",
        "heatmap_df = de_global[['scale1','group1']]\n",
        "heatmap_df['log10scale1']=np.log10(heatmap_df['scale1'])\n",
        "heatmap_df=heatmap_df[['log10scale1','group1']]\n",
        "heatmap_df=heatmap_df.pivot(columns='group1',values='log10scale1')\n",
        "\n",
        "# put the heatmap data in anndata object \n",
        "heatmap_adata=anndata.AnnData(X=heatmap_df.values.T, \n",
        "                              obs=pd.DataFrame(index=heatmap_df.columns.values), \n",
        "                              var=pd.DataFrame(index=heatmap_df.index.values),\n",
        "                              )\n",
        "#rename obs and var to make clear what they hold\n",
        "heatmap_adata.var.index.rename('gene_id',inplace=True)\n",
        "heatmap_adata.obs.index.rename('cell_type',inplace=True)\n",
        "\n",
        "# add some meatadata explaining what the data is\n",
        "heatmap_adata.uns['about']=\"\"\"\n",
        "bendavid2021 heatmap data for 19 cell types\n",
        "this h5ad file is made to be used with WormBase wormcells-viz heatmap:\n",
        "https://github.com/WormBase/wormcells-viz\n",
        "\"\"\"\n",
        "heatmap_adata.write_h5ad('heatmap_data_bendavid2021.h5ad')\n",
        "heatmap_adata"
      ],
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  \"\"\"\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "AnnData object with n_obs × n_vars = 19 × 1094\n",
              "    uns: 'about'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MqW10MedBews"
      },
      "source": [
        "**Make gene expression histogram anndata**\n",
        "\n",
        "This data is a 3D tensor of shape: \n",
        "\n",
        "$ n_{celltypes} \\times n_{bins} \\times n_{genes} = n_{obs} \\times n_{var} \\times n_{layers} $  \n",
        "\n",
        "The anndata obs contains the cell types and var contains the histogram bins, the genes are stored in layers with the keys being the gene ID.\n",
        "\n",
        "We store the genes in the layers because each view in the wormcells-viz app show the histograms for a single gene, so this makes accessing the data simpler\n",
        "\n",
        "The histogram bin counts are computed from the scvi normalized expression values, binned in 100 bins from 10^-9 to 10^0\n",
        "\n",
        "```\n",
        "anndata.obs = cell_types\n",
        "anndata.var = bins of counts seen in range (100 evenly spaced bins, from -10 to 0) because the data is log10 transformed\n",
        "anndata.X = NOTHING (filled with vector of zeroes)\n",
        "anndata.layers[cell_type] = the key is the corresponding cell_type, each layer\n",
        "contains counts in each bin for all cell types\n",
        "anndata.uns['about'] = information about the dataset\n",
        "```\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DFtSrZLa0KcO",
        "outputId": "de9531eb-b4bf-4b0b-8728-7c3ff8faccae"
      },
      "source": [
        "bins_intervals=np.histogram([0], bins=100, range=(-10,0), density=False)[1][:-1]\n",
        "bins_intervals"
      ],
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array([-10. ,  -9.9,  -9.8,  -9.7,  -9.6,  -9.5,  -9.4,  -9.3,  -9.2,\n",
              "        -9.1,  -9. ,  -8.9,  -8.8,  -8.7,  -8.6,  -8.5,  -8.4,  -8.3,\n",
              "        -8.2,  -8.1,  -8. ,  -7.9,  -7.8,  -7.7,  -7.6,  -7.5,  -7.4,\n",
              "        -7.3,  -7.2,  -7.1,  -7. ,  -6.9,  -6.8,  -6.7,  -6.6,  -6.5,\n",
              "        -6.4,  -6.3,  -6.2,  -6.1,  -6. ,  -5.9,  -5.8,  -5.7,  -5.6,\n",
              "        -5.5,  -5.4,  -5.3,  -5.2,  -5.1,  -5. ,  -4.9,  -4.8,  -4.7,\n",
              "        -4.6,  -4.5,  -4.4,  -4.3,  -4.2,  -4.1,  -4. ,  -3.9,  -3.8,\n",
              "        -3.7,  -3.6,  -3.5,  -3.4,  -3.3,  -3.2,  -3.1,  -3. ,  -2.9,\n",
              "        -2.8,  -2.7,  -2.6,  -2.5,  -2.4,  -2.3,  -2.2,  -2.1,  -2. ,\n",
              "        -1.9,  -1.8,  -1.7,  -1.6,  -1.5,  -1.4,  -1.3,  -1.2,  -1.1,\n",
              "        -1. ,  -0.9,  -0.8,  -0.7,  -0.6,  -0.5,  -0.4,  -0.3,  -0.2,\n",
              "        -0.1])"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 12
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FxSJ85FWCx4v",
        "outputId": "f0ee20df-7ee6-4b55-f97c-86683ab0229a"
      },
      "source": [
        "### get the scvi normalized expression then log10 that\n",
        "adata.layers['normalized'] = model.get_normalized_expression()\n",
        "adata.layers['log10normalized']=np.log10(adata.layers['normalized'])\n",
        "\n",
        "###loops through each cell type and then each gene to compute the histogram of expression \n",
        "\n",
        "# first get dimensions to initialize adata object\n",
        "celltypes=adata.obs.cell_type.unique()\n",
        "\n",
        "# gets the bin intervals from the np histogram function\n",
        "nbins=100\n",
        "histogram_range=(-10,0)\n",
        "bin_intervals=np.histogram([0], bins=nbins, range=histogram_range, density=False)[1][:-1]\n",
        "# converts list of bins to string for anndata var index\n",
        "bin_intervals=np.round(list(bins_intervals),1).astype(str)\n",
        "gene_histogram_adata=anndata.AnnData(X=np.zeros((len(celltypes),len(bins_intervals))), \n",
        "                              var=pd.DataFrame(index=bin_intervals), \n",
        "                              obs=pd.DataFrame(index=celltypes),\n",
        "                              )\n",
        "#rename obs and var to make clear what they hold\n",
        "gene_histogram_adata.var.index.rename('histogram_bins',inplace=True)\n",
        "gene_histogram_adata.obs.index.rename('cell_type',inplace=True)\n",
        "\n",
        "# now that adata is ready loop through every gene\n",
        "# and for each gene computes the counts in each bin for each cell type\n",
        "for gene_id in tqdm(adata.var.index): \n",
        "    log10_normalized_expression_in_gene=adata[:,adata.var.index==gene_id].layers['log10normalized']\n",
        "    log10_normalized_expression_in_gene=np.squeeze(np.asarray(log10_normalized_expression_in_gene)) \n",
        "    # gets the bin intervals from the np histogram function\n",
        "    gene_histogram_df=pd.DataFrame(columns=bins_intervals)\n",
        "    for cell_type in adata.obs['cell_type'].unique():\n",
        "        # fetch only the expression of that gene in that cell\n",
        "        log10_normalized_expression_in_celltype=log10_normalized_expression_in_gene[adata.obs['cell_type']==cell_type]                                                    \n",
        "        gene_histogram_df.loc[cell_type]=np.histogram(log10_normalized_expression_in_celltype, bins=100, range=(-10,0), density=False)[0]\n",
        "        gene_histogram_df.loc[cell_type]=np.histogram(log10_normalized_expression_in_celltype, bins=100, range=(-10,0), density=False)[0]\n",
        "    # need to convert these to int or anndata doesn't save    \n",
        "    gene_histogram_adata.layers[gene_id]=gene_histogram_df.values.astype(int)\n",
        "\n",
        "# add some meatadata explaining what the data is\n",
        "gene_histogram_adata.uns['about']=\"\"\"\n",
        "bendavid2021 histogram data for 19 cell types and 1k genes\n",
        "this h5ad file is made to be used with WormBase wormcells-viz histogram:\n",
        "https://github.com/WormBase/wormcells-viz\n",
        "\"\"\"\n",
        "gene_histogram_adata.write_h5ad('histogram_data_bendavid2021.h5ad')\n",
        "gene_histogram_adata"
      ],
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/anndata/_core/anndata.py:120: ImplicitModificationWarning: Transforming to str index.\n",
            "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
            "100%|██████████| 1094/1094 [01:32<00:00, 11.81it/s]\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "AnnData object with n_obs × n_vars = 19 × 100\n",
              "    uns: 'about'\n",
              "    layers: 'WBGene00010957', 'WBGene00010958', 'WBGene00010959', 'WBGene00010960', 'WBGene00010961', 'WBGene00000829', 'WBGene00010962', 'WBGene00010963', 'WBGene00010964', 'WBGene00010965', 'WBGene00010967', 'WBGene00017926', 'WBGene00001071', 'WBGene00003932', 'WBGene00022048', 'WBGene00004493', 'WBGene00020297', 'WBGene00022046', 'WBGene00019772', 'WBGene00004427', 'WBGene00000677', 'WBGene00003893', 'WBGene00000681', 'WBGene00004497', 'WBGene00003947', 'WBGene00019537', 'WBGene00019540', 'WBGene00016918', 'WBGene00006920', 'WBGene00021427', 'WBGene00021420', 'WBGene00022089', 'WBGene00022093', 'WBGene00021875', 'WBGene00021894', 'WBGene00021883', 'WBGene00021888', 'WBGene00022069', 'WBGene00022072', 'WBGene00003153', 'WBGene00021088', 'WBGene00001999', 'WBGene00015248', 'WBGene00015413', 'WBGene00001702', 'WBGene00001692', 'WBGene00001699', 'WBGene00003955', 'WBGene00001074', 'WBGene00000611', 'WBGene00018898', 'WBGene00004419', 'WBGene00019947', 'WBGene00017925', 'WBGene00020679', 'WBGene00017119', 'WBGene00017121', 'WBGene00004432', 'WBGene00019162', 'WBGene00006586', 'WBGene00021161', 'WBGene00001244', 'WBGene00020585', 'WBGene00016289', 'WBGene00004494', 'WBGene00006780', 'WBGene00016722', 'WBGene00004112', 'WBGene00001853', 'WBGene00022235', 'WBGene00022254', 'WBGene00022245', 'WBGene00001029', 'WBGene00001858', 'WBGene00006917', 'WBGene00020683', 'WBGene00003992', 'WBGene00004473', 'WBGene00000942', 'WBGene00004477', 'WBGene00016739', 'WBGene00019900', 'WBGene00017991', 'WBGene00017993', 'WBGene00018468', 'WBGene00004046', 'WBGene00003758', 'WBGene00015619', 'WBGene00017998', 'WBGene00004703', 'WBGene00004363', 'WBGene00016790', 'WBGene00004471', 'WBGene00000183', 'WBGene00003178', 'WBGene00018361', 'WBGene00002162', 'WBGene00019510', 'WBGene00017075', 'WBGene00004995', 'WBGene00000691', 'WBGene00000594', 'WBGene00000692', 'WBGene00001132', 'WBGene00011708', 'WBGene00000266', 'WBGene00013514', 'WBGene00013882', 'WBGene00007223', 'WBGene00001444', 'WBGene00012437', 'WBGene00004463', 'WBGene00009880', 'WBGene00009882', 'WBGene00010317', 'WBGene00007880', 'WBGene00001239', 'WBGene00011743', 'WBGene00010639', 'WBGene00001113', 'WBGene00008149', 'WBGene00005663', 'WBGene00010677', 'WBGene00010556', 'WBGene00001794', 'WBGene00007696', 'WBGene00008506', 'WBGene00002152', 'WBGene00001189', 'WBGene00001328', 'WBGene00011289', 'WBGene00014006', 'WBGene00014003', 'WBGene00013477', 'WBGene00001426', 'WBGene00012382', 'WBGene00012383', 'WBGene00002344', 'WBGene00004474', 'WBGene00011481', 'WBGene00011480', 'WBGene00004204', 'WBGene00008505', 'WBGene00007443', 'WBGene00000699', 'WBGene00007107', 'WBGene00001938', 'WBGene00003148', 'WBGene00004480', 'WBGene00003415', 'WBGene00014088', 'WBGene00014086', 'WBGene00003920', 'WBGene00014091', 'WBGene00014095', 'WBGene00006759', 'WBGene00000122', 'WBGene00004448', 'WBGene00004492', 'WBGene00004315', 'WBGene00001452', 'WBGene00008554', 'WBGene00007868', 'WBGene00009896', 'WBGene00008028', 'WBGene00004424', 'WBGene00010440', 'WBGene00044737', 'WBGene00013463', 'WBGene00001393', 'WBGene00004410', 'WBGene00003755', 'WBGene00012875', 'WBGene00004430', 'WBGene00004443', 'WBGene00003085', 'WBGene00009926', 'WBGene00011019', 'WBGene00010794', 'WBGene00013289', 'WBGene00001558', 'WBGene00013308', 'WBGene00013311', 'WBGene00004487', 'WBGene00001066', 'WBGene00012768', 'WBGene00000370', 'WBGene00001604', 'WBGene00004357', 'WBGene00013784', 'WBGene00004914', 'WBGene00002005', 'WBGene00011374', 'WBGene00008707', 'WBGene00008706', 'WBGene00021043', 'WBGene00020703', 'WBGene00002269', 'WBGene00020561', 'WBGene00016028', 'WBGene00000882', 'WBGene00004498', 'WBGene00019327', 'WBGene00019332', 'WBGene00001392', 'WBGene00000263', 'WBGene00022497', 'WBGene00022336', 'WBGene00022053', 'WBGene00021248', 'WBGene00002025', 'WBGene00000941', 'WBGene00004491', 'WBGene00021857', 'WBGene00021849', 'WBGene00022194', 'WBGene00022170', 'WBGene00007000', 'WBGene00006725', 'WBGene00019168', 'WBGene00003175', 'WBGene00006452', 'WBGene00021800', 'WBGene00003372', 'WBGene00010303', 'WBGene00007924', 'WBGene00004428', 'WBGene00004451', 'WBGene00003962', 'WBGene00002267', 'WBGene00004414', 'WBGene00003930', 'WBGene00014204', 'WBGene00014207', 'WBGene00003748', 'WBGene00006537', 'WBGene00002202', 'WBGene00007630', 'WBGene00003073', 'WBGene00006959', 'WBGene00004387', 'WBGene00004470', 'WBGene00002100', 'WBGene00000380', 'WBGene00003915', 'WBGene00009440', 'WBGene00000206', 'WBGene00011408', 'WBGene00004469', 'WBGene00020097', 'WBGene00016677', 'WBGene00006519', 'WBGene00021899', 'WBGene00003588', 'WBGene00000229', 'WBGene00022599', 'WBGene00003901', 'WBGene00019831', 'WBGene00003371', 'WBGene00018963', 'WBGene00004481', 'WBGene00004435', 'WBGene00015148', 'WBGene00000182', 'WBGene00000381', 'WBGene00017166', 'WBGene00016194', 'WBGene00006727', 'WBGene00003375', 'WBGene00021532', 'WBGene00018341', 'WBGene00004472', 'WBGene00016977', 'WBGene00004482', 'WBGene00015811', 'WBGene00002850', 'WBGene00001005', 'WBGene00003162', 'WBGene00000998', 'WBGene00020317', 'WBGene00002083', 'WBGene00004133', 'WBGene00001168', 'WBGene00015103', 'WBGene00004483', 'WBGene00004450', 'WBGene00004490', 'WBGene00020112', 'WBGene00004417', 'WBGene00020636', 'WBGene00000118', 'WBGene00004420', 'WBGene00003063', 'WBGene00001209', 'WBGene00004449', 'WBGene00004918', 'WBGene00000845', 'WBGene00000665', 'WBGene00018846', 'WBGene00000041', 'WBGene00018407', 'WBGene00019680', 'WBGene00004433', 'WBGene00015755', 'WBGene00002011', 'WBGene00004187', 'WBGene00000414', 'WBGene00016250', 'WBGene00004460', 'WBGene00003929', 'WBGene00014030', 'WBGene00001227', 'WBGene00011300', 'WBGene00001385', 'WBGene00001386', 'WBGene00001263', 'WBGene00004980', 'WBGene00000235', 'WBGene00001094', 'WBGene00000550', 'WBGene00006704', 'WBGene00001570', 'WBGene00011498', 'WBGene00002064', 'WBGene00001155', 'WBGene00006910', 'WBGene00006911', 'WBGene00000567', 'WBGene00010539', 'WBGene00010541', 'WBGene00003482', 'WBGene00011869', 'WBGene00011870', 'WBGene00000833', 'WBGene00003225', 'WBGene00011561', 'WBGene00004798', 'WBGene00012361', 'WBGene00006536', 'WBGene00004302', 'WBGene00010478', 'WBGene00004736', 'WBGene00000475', 'WBGene00000667', 'WBGene00000668', 'WBGene00000669', 'WBGene00012964', 'WBGene00003759', 'WBGene00012760', 'WBGene00012759', 'WBGene00013238', 'WBGene00000273', 'WBGene00001946', 'WBGene00004919', 'WBGene00004700', 'WBGene00012544', 'WBGene00012550', 'WBGene00012553', 'WBGene00001457', 'WBGene00006762', 'WBGene00006728', 'WBGene00000673', 'WBGene00010139', 'WBGene00002059', 'WBGene00009732', 'WBGene00012097', 'WBGene00012094', 'WBGene00006439', 'WBGene00011474', 'WBGene00000067', 'WBGene00011736', 'WBGene00011730', 'WBGene00020216', 'WBGene00003991', 'WBGene00018656', 'WBGene00001065', 'WBGene00004423', 'WBGene00020194', 'WBGene00020195', 'WBGene00003776', 'WBGene00001461', 'WBGene00000175', 'WBGene00018701', 'WBGene00001685', 'WBGene00001684', 'WBGene00004951', 'WBGene00002255', 'WBGene00002253', 'WBGene00002254', 'WBGene00006863', 'WBGene00015954', 'WBGene00015354', 'WBGene00015356', 'WBGene00006408', 'WBGene00018532', 'WBGene00018025', 'WBGene00018024', 'WBGene00015823', 'WBGene00000207', 'WBGene00003963', 'WBGene00001455', 'WBGene00001454', 'WBGene00004438', 'WBGene00018877', 'WBGene00000066', 'WBGene00019759', 'WBGene00019760', 'WBGene00022500', 'WBGene00022499', 'WBGene00006861', 'WBGene00000149', 'WBGene00001070', 'WBGene00006983', 'WBGene00016133', 'WBGene00044623', 'WBGene00006826', 'WBGene00019160', 'WBGene00002023', 'WBGene00004219', 'WBGene00002007', 'WBGene00015805', 'WBGene00015803', 'WBGene00015781', 'WBGene00015778', 'WBGene00000739', 'WBGene00000740', 'WBGene00000741', 'WBGene00000742', 'WBGene00003495', 'WBGene00017691', 'WBGene00020886', 'WBGene00000210', 'WBGene00015168', 'WBGene00018519', 'WBGene00018522', 'WBGene00004990', 'WBGene00001069', 'WBGene00017864', 'WBGene00016898', 'WBGene00004999', 'WBGene00003171', 'WBGene00001169', 'WBGene00006588', 'WBGene00001945', 'WBGene00001000', 'WBGene00019780', 'WBGene00006942', 'WBGene00001448', 'WBGene00006587', 'WBGene00017149', 'WBGene00019727', 'WBGene00006764', 'WBGene00003752', 'WBGene00006672', 'WBGene00016033', 'WBGene00000100', 'WBGene00016006', 'WBGene00006534', 'WBGene00009888', 'WBGene00001863', 'WBGene00002271', 'WBGene00000749', 'WBGene00044686', 'WBGene00001331', 'WBGene00008603', 'WBGene00009621', 'WBGene00001462', 'WBGene00009813', 'WBGene00011880', 'WBGene00012261', 'WBGene00012255', 'WBGene00012256', 'WBGene00012257', 'WBGene00009531', 'WBGene00006584', 'WBGene00008572', 'WBGene00001502', 'WBGene00008205', 'WBGene00012158', 'WBGene00000277', 'WBGene00009634', 'WBGene00009640', 'WBGene00009637', 'WBGene00009638', 'WBGene00009639', 'WBGene00003514', 'WBGene00010424', 'WBGene00011771', 'WBGene00000217', 'WBGene00003903', 'WBGene00007848', 'WBGene00010994', 'WBGene00010940', 'WBGene00001898', 'WBGene00008357', 'WBGene00008359', 'WBGene00008358', 'WBGene00008360', 'WBGene00008356', 'WBGene00006366', 'WBGene00000822', 'WBGene00006634', 'WBGene00050939', 'WBGene00007350', 'WBGene00008605', 'WBGene00011683', 'WBGene00009787', 'WBGene00010337', 'WBGene00003169', 'WBGene00009069', 'WBGene00044638', 'WBGene00002280', 'WBGene00018187', 'WBGene00045272', 'WBGene00017727', 'WBGene00020143', 'WBGene00006886', 'WBGene00017376', 'WBGene00003891', 'WBGene00003370', 'WBGene00003369', 'WBGene00002272', 'WBGene00003744', 'WBGene00020732', 'WBGene00022276', 'WBGene00004274', 'WBGene00004418', 'WBGene00002077', 'WBGene00004143', 'WBGene00021026', 'WBGene00022042', 'WBGene00001725', 'WBGene00004429', 'WBGene00004888', 'WBGene00022271', 'WBGene00006771', 'WBGene00022144', 'WBGene00016907', 'WBGene00004441', 'WBGene00001196', 'WBGene00004409', 'WBGene00021350', 'WBGene00004014', 'WBGene00021466', 'WBGene00001404', 'WBGene00004920', 'WBGene00022127', 'WBGene00021128', 'WBGene00004412', 'WBGene00022114', 'WBGene00022122', 'WBGene00020915', 'WBGene00021036', 'WBGene00019361', 'WBGene00001501', 'WBGene00002135', 'WBGene00002134', 'WBGene00004425', 'WBGene00006833', 'WBGene00020192', 'WBGene00001430', 'WBGene00003950', 'WBGene00000779', 'WBGene00006820', 'WBGene00004464', 'WBGene00000140', 'WBGene00003061', 'WBGene00000962', 'WBGene00004113', 'WBGene00004479', 'WBGene00004436', 'WBGene00016653', 'WBGene00004415', 'WBGene00006961', 'WBGene00019767', 'WBGene00023067', 'WBGene00023068', 'WBGene00006984', 'WBGene00017671', 'WBGene00003934', 'WBGene00016733', 'WBGene00003924', 'WBGene00004094', 'WBGene00004131', 'WBGene00000230', 'WBGene00001067', 'WBGene00004431', 'WBGene00000474', 'WBGene00006919', 'WBGene00000473', 'WBGene00018893', 'WBGene00001856', 'WBGene00016422', 'WBGene00022748', 'WBGene00020382', 'WBGene00004486', 'WBGene00005000', 'WBGene00015687', 'WBGene00016943', 'WBGene00020417', 'WBGene00000896', 'WBGene00019001', 'WBGene00017065', 'WBGene00002258', 'WBGene00001682', 'WBGene00006819', 'WBGene00016493', 'WBGene00017098', 'WBGene00019324', 'WBGene00019322', 'WBGene00001598', 'WBGene00006423', 'WBGene00010370', 'WBGene00003964', 'WBGene00012166', 'WBGene00009051', 'WBGene00009056', 'WBGene00010738', 'WBGene00011061', 'WBGene00009004', 'WBGene00008765', 'WBGene00009915', 'WBGene00006754', 'WBGene00012140', 'WBGene00001648', 'WBGene00008414', 'WBGene00009189', 'WBGene00011232', 'WBGene00011917', 'WBGene00000869', 'WBGene00010560', 'WBGene00013958', 'WBGene00009920', 'WBGene00004439', 'WBGene00003596', 'WBGene00009664', 'WBGene00000777', 'WBGene00009562', 'WBGene00000209', 'WBGene00003953', 'WBGene00003925', 'WBGene00006574', 'WBGene00003052', 'WBGene00009454', 'WBGene00004484', 'WBGene00010135', 'WBGene00011629', 'WBGene00009958', 'WBGene00009342', 'WBGene00012376', 'WBGene00001167', 'WBGene00009126', 'WBGene00002026', 'WBGene00004808', 'WBGene00004807', 'WBGene00077771', 'WBGene00004488', 'WBGene00004437', 'WBGene00004426', 'WBGene00001145', 'WBGene00009161', 'WBGene00006528', 'WBGene00000371', 'WBGene00009408', 'WBGene00006978', 'WBGene00006633', 'WBGene00001226', 'WBGene00007155', 'WBGene00077526', 'WBGene00000931', 'WBGene00013859', 'WBGene00004476', 'WBGene00001337', 'WBGene00003902', 'WBGene00004444', 'WBGene00009119', 'WBGene00003926', 'WBGene00004408', 'WBGene00009122', 'WBGene00044070', 'WBGene00015232', 'WBGene00016393', 'WBGene00002191', 'WBGene00018849', 'WBGene00013181', 'WBGene00014183', 'WBGene00008857', 'WBGene00006876', 'WBGene00011155', 'WBGene00011156', 'WBGene00003123', 'WBGene00011059', 'WBGene00002000', 'WBGene00011026', 'WBGene00012216', 'WBGene00011586', 'WBGene00010419', 'WBGene00011548', 'WBGene00006889', 'WBGene00012484', 'WBGene00001427', 'WBGene00003989', 'WBGene00006529', 'WBGene00003948', 'WBGene00001234', 'WBGene00003753', 'WBGene00006595', 'WBGene00004915', 'WBGene00012344', 'WBGene00007219', 'WBGene00012289', 'WBGene00012738', 'WBGene00012351', 'WBGene00004445', 'WBGene00012354', 'WBGene00004475', 'WBGene00008191', 'WBGene00006824', 'WBGene00004489', 'WBGene00002978', 'WBGene00001030', 'WBGene00004495', 'WBGene00006789', 'WBGene00006845', 'WBGene00001488', 'WBGene00002189', 'WBGene00009367', 'WBGene00000881', 'WBGene00004496', 'WBGene00021981', 'WBGene00001564', 'WBGene00019983', 'WBGene00019563', 'WBGene00020155', 'WBGene00000552', 'WBGene00004266', 'WBGene00022287', 'WBGene00006794', 'WBGene00002016', 'WBGene00002018', 'WBGene00020391', 'WBGene00019092', 'WBGene00021486', 'WBGene00001758', 'WBGene00004130', 'WBGene00016855', 'WBGene00016483', 'WBGene00020869', 'WBGene00015433', 'WBGene00020696', 'WBGene00004422', 'WBGene00020854', 'WBGene00016781', 'WBGene00020237', 'WBGene00003747', 'WBGene00003751', 'WBGene00017982', 'WBGene00017984', 'WBGene00021625', 'WBGene00018811', 'WBGene00018297', 'WBGene00019357', 'WBGene00001855', 'WBGene00016509', 'WBGene00020590', 'WBGene00020589', 'WBGene00020588', 'WBGene00015340', 'WBGene00015339', 'WBGene00001037', 'WBGene00003927', 'WBGene00002637', 'WBGene00004502', 'WBGene00000534', 'WBGene00000802', 'WBGene00006436', 'WBGene00021068', 'WBGene00020275', 'WBGene00020269', 'WBGene00002273', 'WBGene00001073', 'WBGene00018491', 'WBGene00004897', 'WBGene00020950', 'WBGene00020808', 'WBGene00006803', 'WBGene00006801', 'WBGene00016362', 'WBGene00017613', 'WBGene00017621', 'WBGene00021959', 'WBGene00017850', 'WBGene00017317', 'WBGene00017363', 'WBGene00017362', 'WBGene00015393', 'WBGene00020054', 'WBGene00017600', 'WBGene00017605', 'WBGene00017201', 'WBGene00015706', 'WBGene00022410', 'WBGene00020246', 'WBGene00004453', 'WBGene00004499', 'WBGene00001464', 'WBGene00003242', 'WBGene00016496', 'WBGene00018221', 'WBGene00017018', 'WBGene00000219', 'WBGene00018692', 'WBGene00020722', 'WBGene00007055', 'WBGene00020862', 'WBGene00002019', 'WBGene00002017', 'WBGene00017657', 'WBGene00000717', 'WBGene00000599', 'WBGene00019466', 'WBGene00001449', 'WBGene00010593', 'WBGene00001909', 'WBGene00001915', 'WBGene00001910', 'WBGene00001911', 'WBGene00010727', 'WBGene00000720', 'WBGene00009772', 'WBGene00007784', 'WBGene00007785', 'WBGene00001172', 'WBGene00014108', 'WBGene00007541', 'WBGene00004364', 'WBGene00000722', 'WBGene00010627', 'WBGene00009340', 'WBGene00003065', 'WBGene00003923', 'WBGene00001303', 'WBGene00008914', 'WBGene00000726', 'WBGene00008920', 'WBGene00011428', 'WBGene00000065', 'WBGene00000064', 'WBGene00000063', 'WBGene00007385', 'WBGene00003771', 'WBGene00010742', 'WBGene00011831', 'WBGene00011383', 'WBGene00000554', 'WBGene00009388', 'WBGene00003762', 'WBGene00014826', 'WBGene00010790', 'WBGene00010290', 'WBGene00004917', 'WBGene00001088', 'WBGene00003057', 'WBGene00003515', 'WBGene00011015', 'WBGene00009092', 'WBGene00005018', 'WBGene00009096', 'WBGene00009097', 'WBGene00008298', 'WBGene00000228', 'WBGene00013853', 'WBGene00010174', 'WBGene00012276', 'WBGene00011244', 'WBGene00004442', 'WBGene00006512', 'WBGene00001137', 'WBGene00000733', 'WBGene00009982', 'WBGene00009983', 'WBGene00001479', 'WBGene00009995', 'WBGene00009993', 'WBGene00009992', 'WBGene00044921', 'WBGene00008628', 'WBGene00010260', 'WBGene00008395', 'WBGene00010513', 'WBGene00010515', 'WBGene00010514', 'WBGene00010516', 'WBGene00013025', 'WBGene00000877', 'WBGene00007617', 'WBGene00008262', 'WBGene00000915', 'WBGene00006836', 'WBGene00012291', 'WBGene00004336', 'WBGene00013884', 'WBGene00004485', 'WBGene00001423', 'WBGene00007235', 'WBGene00000883', 'WBGene00000879', 'WBGene00004452', 'WBGene00000776', 'WBGene00006585', 'WBGene00013077', 'WBGene00013078', 'WBGene00013079', 'WBGene00013081', 'WBGene00013489', 'WBGene00013481', 'WBGene00013588', 'WBGene00000214', 'WBGene00012813', 'WBGene00012812', 'WBGene00000920', 'WBGene00013766', 'WBGene00004413', 'WBGene00012615', 'WBGene00012186', 'WBGene00012853', 'WBGene00011392', 'WBGene00001431', 'WBGene00006812', 'WBGene00005017', 'WBGene00004504', 'WBGene00000857', 'WBGene00020149', 'WBGene00044640', 'WBGene00019282', 'WBGene00021073', 'WBGene00020836', 'WBGene00021095', 'WBGene00020181', 'WBGene00018572', 'WBGene00006924', 'WBGene00044301', 'WBGene00020229', 'WBGene00015913', 'WBGene00019488', 'WBGene00004135', 'WBGene00004467', 'WBGene00000928', 'WBGene00001976', 'WBGene00017374', 'WBGene00018271', 'WBGene00000606', 'WBGene00001755', 'WBGene00019893', 'WBGene00022765', 'WBGene00003949', 'WBGene00020868', 'WBGene00004434', 'WBGene00044389', 'WBGene00006434', 'WBGene00003097', 'WBGene00015920', 'WBGene00016118', 'WBGene00015943', 'WBGene00004267', 'WBGene00002054', 'WBGene00002053', 'WBGene00002268', 'WBGene00018869', 'WBGene00000403', 'WBGene00004015', 'WBGene00017440', 'WBGene00016210', 'WBGene00016316', 'WBGene00002980', 'WBGene00019061', 'WBGene00004916', 'WBGene00004702', 'WBGene00020894', 'WBGene00007050', 'WBGene00002045', 'WBGene00020581', 'WBGene00001064', 'WBGene00018303', 'WBGene00018298', 'WBGene00004921', 'WBGene00016630', 'WBGene00022856', 'WBGene00002238', 'WBGene00004447', 'WBGene00017347', 'WBGene00016173', 'WBGene00017127', 'WBGene00017970', 'WBGene00002008', 'WBGene00018395', 'WBGene00006956', 'WBGene00004398', 'WBGene00019940', 'WBGene00020114', 'WBGene00015061', 'WBGene00015059', 'WBGene00015514', 'WBGene00011638', 'WBGene00006724', 'WBGene00004421', 'WBGene00000377', 'WBGene00004264', 'WBGene00002066', 'WBGene00003821', 'WBGene00000379', 'WBGene00001840', 'WBGene00002065', 'WBGene00004416', 'WBGene00010044', 'WBGene00004440', 'WBGene00004397', 'WBGene00000240', 'WBGene00000378', 'WBGene00011884', 'WBGene00007684', 'WBGene00007683', 'WBGene00007686', 'WBGene00004446', 'WBGene00011977', 'WBGene00000170', 'WBGene00006433', 'WBGene00010942', 'WBGene00003749', 'WBGene00007463', 'WBGene00007331', 'WBGene00007330', 'WBGene00011273', 'WBGene00002265', 'WBGene00001447', 'WBGene00014165', 'WBGene00011038', 'WBGene00006918', 'WBGene00008480', 'WBGene00006583', 'WBGene00010321', 'WBGene00007355', 'WBGene00010896', 'WBGene00000616', 'WBGene00011072', 'WBGene00011070', 'WBGene00003367', 'WBGene00001753', 'WBGene00004300', 'WBGene00003756', 'WBGene00000472', 'WBGene00009353', 'WBGene00004454', 'WBGene00003831', 'WBGene00004478', 'WBGene00009578', 'WBGene00009577', 'WBGene00009514', 'WBGene00005016', 'WBGene00007192', 'WBGene00004698', 'WBGene00007142', 'WBGene00013387', 'WBGene00001031', 'WBGene00008549', 'WBGene00006617', 'WBGene00012148', 'WBGene00002262', 'WBGene00000114', 'WBGene00001679', 'WBGene00009691', 'WBGene00009692', 'WBGene00009688', 'WBGene00001977', 'WBGene00006601', 'WBGene00004167', 'WBGene00013263', 'WBGene00008841', 'WBGene00006999', 'WBGene00002264', 'WBGene00001459', 'WBGene00008446', 'WBGene00009740', 'WBGene00013007', 'WBGene00013094', 'WBGene00010723', 'WBGene00012978', 'WBGene00004456', 'WBGene00001971', 'WBGene00009180', 'WBGene00008963', 'WBGene00006787', 'WBGene00001225', 'WBGene00012179'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ESjbD3ejjPzt"
      },
      "source": [
        "###  Compute pairwise DE between all cell types for generating data for swarm plots"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "collapsed": true,
        "jupyter": {
          "outputs_hidden": true
        },
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 439
        },
        "id": "MdU-DTTpfC5x",
        "outputId": "9eaca2e1-323d-4d1b-b926-fa5c56b79f6a"
      },
      "source": [
        "# do pairwise DE on all the 19 top level cell types defined in bendavid2021\n",
        "celltypes=adata.obs.cell_type.unique()\n",
        "\n",
        "# the code below only does 500 samples to compute DE quickly to make test data\n",
        "# default is 5000 epochs, should be used when generating data for a deploy\n",
        "pairwise_celltypes=list(itertools.product(celltypes, celltypes))\n",
        "pairwise_de=pd.DataFrame()\n",
        "for cellpair in tqdm(pairwise_celltypes):\n",
        "    celltype1=cellpair[0]\n",
        "    celltype2=cellpair[1]\n",
        "    de_df = model.differential_expression(\n",
        "    groupby='cell_type',\n",
        "    group1=celltype1,\n",
        "    group2=celltype2,\n",
        "    silent=True,\n",
        "    n_samples=500, # ONLY USING 500 SAMPLES IS TOO LITTLE, I DO IT FOR SPEED IN THIS EXAMPLE\n",
        "    all_stats=False\n",
        "    )\n",
        "    de_df['group1']=celltype1\n",
        "    de_df['group2']=celltype2\n",
        "    pairwise_de=pairwise_de.append(de_df)\n",
        "#write to disk just in case\n",
        "pairwise_de.to_csv('bendavid2021_pairwisede_19cell_types.csv')\n",
        "pairwise_de"
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "100%|██████████| 361/361 [04:16<00:00,  1.41it/s]\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>proba_de</th>\n",
              "      <th>proba_not_de</th>\n",
              "      <th>bayes_factor</th>\n",
              "      <th>scale1</th>\n",
              "      <th>scale2</th>\n",
              "      <th>pseudocounts</th>\n",
              "      <th>delta</th>\n",
              "      <th>lfc_mean</th>\n",
              "      <th>lfc_median</th>\n",
              "      <th>lfc_std</th>\n",
              "      <th>lfc_min</th>\n",
              "      <th>lfc_max</th>\n",
              "      <th>is_de_fdr_0.05</th>\n",
              "      <th>comparison</th>\n",
              "      <th>group1</th>\n",
              "      <th>group2</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>WBGene00000065</th>\n",
              "      <td>0.968</td>\n",
              "      <td>0.032</td>\n",
              "      <td>3.409496</td>\n",
              "      <td>0.000688</td>\n",
              "      <td>0.000722</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>-0.065180</td>\n",
              "      <td>-0.054506</td>\n",
              "      <td>3.274759</td>\n",
              "      <td>-11.658871</td>\n",
              "      <td>11.782922</td>\n",
              "      <td>True</td>\n",
              "      <td>Intestine vs Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00006819</th>\n",
              "      <td>0.962</td>\n",
              "      <td>0.038</td>\n",
              "      <td>3.231428</td>\n",
              "      <td>0.000361</td>\n",
              "      <td>0.000408</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>-0.112533</td>\n",
              "      <td>-0.026206</td>\n",
              "      <td>3.667365</td>\n",
              "      <td>-13.391572</td>\n",
              "      <td>12.311786</td>\n",
              "      <td>True</td>\n",
              "      <td>Intestine vs Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00003495</th>\n",
              "      <td>0.960</td>\n",
              "      <td>0.040</td>\n",
              "      <td>3.178054</td>\n",
              "      <td>0.000478</td>\n",
              "      <td>0.000471</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.026722</td>\n",
              "      <td>-0.138340</td>\n",
              "      <td>4.723500</td>\n",
              "      <td>-15.807770</td>\n",
              "      <td>16.493458</td>\n",
              "      <td>True</td>\n",
              "      <td>Intestine vs Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00003934</th>\n",
              "      <td>0.958</td>\n",
              "      <td>0.042</td>\n",
              "      <td>3.127178</td>\n",
              "      <td>0.004401</td>\n",
              "      <td>0.004435</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>-0.034704</td>\n",
              "      <td>-0.133943</td>\n",
              "      <td>6.269757</td>\n",
              "      <td>-21.310509</td>\n",
              "      <td>23.077679</td>\n",
              "      <td>True</td>\n",
              "      <td>Intestine vs Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00013784</th>\n",
              "      <td>0.956</td>\n",
              "      <td>0.044</td>\n",
              "      <td>3.078568</td>\n",
              "      <td>0.000011</td>\n",
              "      <td>0.000011</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>-0.014021</td>\n",
              "      <td>0.180369</td>\n",
              "      <td>3.034412</td>\n",
              "      <td>-11.592129</td>\n",
              "      <td>11.813314</td>\n",
              "      <td>True</td>\n",
              "      <td>Intestine vs Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00004486</th>\n",
              "      <td>0.594</td>\n",
              "      <td>0.406</td>\n",
              "      <td>0.380526</td>\n",
              "      <td>0.004717</td>\n",
              "      <td>0.004599</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.038043</td>\n",
              "      <td>0.041127</td>\n",
              "      <td>0.608247</td>\n",
              "      <td>-1.850594</td>\n",
              "      <td>3.647875</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs XXX</td>\n",
              "      <td>XXX</td>\n",
              "      <td>XXX</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00004410</th>\n",
              "      <td>0.584</td>\n",
              "      <td>0.416</td>\n",
              "      <td>0.339216</td>\n",
              "      <td>0.003237</td>\n",
              "      <td>0.003185</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.023904</td>\n",
              "      <td>0.016881</td>\n",
              "      <td>0.544974</td>\n",
              "      <td>-1.818641</td>\n",
              "      <td>3.619397</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs XXX</td>\n",
              "      <td>XXX</td>\n",
              "      <td>XXX</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00004485</th>\n",
              "      <td>0.568</td>\n",
              "      <td>0.432</td>\n",
              "      <td>0.273696</td>\n",
              "      <td>0.002235</td>\n",
              "      <td>0.002191</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.034557</td>\n",
              "      <td>0.042763</td>\n",
              "      <td>0.562306</td>\n",
              "      <td>-1.818583</td>\n",
              "      <td>4.086172</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs XXX</td>\n",
              "      <td>XXX</td>\n",
              "      <td>XXX</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00004435</th>\n",
              "      <td>0.544</td>\n",
              "      <td>0.456</td>\n",
              "      <td>0.176456</td>\n",
              "      <td>0.005193</td>\n",
              "      <td>0.005098</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>0.028754</td>\n",
              "      <td>0.022504</td>\n",
              "      <td>0.478892</td>\n",
              "      <td>-1.651722</td>\n",
              "      <td>2.787595</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs XXX</td>\n",
              "      <td>XXX</td>\n",
              "      <td>XXX</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00004438</th>\n",
              "      <td>0.540</td>\n",
              "      <td>0.460</td>\n",
              "      <td>0.160343</td>\n",
              "      <td>0.001176</td>\n",
              "      <td>0.001179</td>\n",
              "      <td>0.0</td>\n",
              "      <td>0.25</td>\n",
              "      <td>-0.002779</td>\n",
              "      <td>-0.008003</td>\n",
              "      <td>0.463970</td>\n",
              "      <td>-1.430840</td>\n",
              "      <td>1.858170</td>\n",
              "      <td>False</td>\n",
              "      <td>XXX vs XXX</td>\n",
              "      <td>XXX</td>\n",
              "      <td>XXX</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>394934 rows × 16 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                proba_de  proba_not_de  ...     group1     group2\n",
              "WBGene00000065     0.968         0.032  ...  Intestine  Intestine\n",
              "WBGene00006819     0.962         0.038  ...  Intestine  Intestine\n",
              "WBGene00003495     0.960         0.040  ...  Intestine  Intestine\n",
              "WBGene00003934     0.958         0.042  ...  Intestine  Intestine\n",
              "WBGene00013784     0.956         0.044  ...  Intestine  Intestine\n",
              "...                  ...           ...  ...        ...        ...\n",
              "WBGene00004486     0.594         0.406  ...        XXX        XXX\n",
              "WBGene00004410     0.584         0.416  ...        XXX        XXX\n",
              "WBGene00004485     0.568         0.432  ...        XXX        XXX\n",
              "WBGene00004435     0.544         0.456  ...        XXX        XXX\n",
              "WBGene00004438     0.540         0.460  ...        XXX        XXX\n",
              "\n",
              "[394934 rows x 16 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VrzgladOBSs4"
      },
      "source": [
        "**Make swarm anndata**\n",
        "\n",
        "This data is a 3D tensor of shape: \n",
        "\n",
        "$ x_{celltypes} \\times y_{genes} \\times z_{celltypes} \\times n_{genes} = x_{obs} \\times y_{var} \\times z_{layers} $ \n",
        "\n",
        "Notice that the cell types are repeated along two dimensions, because this data contains the results of pairwise DE comparisons among each cell type in the data.\n",
        "\n",
        "**Plus** # of cell types matrices of shape:\n",
        "$ x_{celltypes} \\times y_{genes} = x_{obs} \\times y_{var}  $\n",
        "\n",
        "There go under `anndata.uns[celltype]`, contains a dataframe with global differential expression results for that celltype, e.g. Intestine vs all other cells.\n",
        "\n",
        "Finally, `anndata.uns['heatmap']` contains the 2D matrix with log10 scvi expression frequency heatmap data, with genes in the index and cell types in the columns. This can be used to display the expression of each tissue upon mouseover.\n",
        "\n",
        "```\n",
        "anndata.obs = cell_types\n",
        "anndata.var = gene_id\n",
        "anndata.X = NOTHING (filled with vector of zeroes)\n",
        "anndata.layers[cell_type] = mean log fold change for a given cell type for all genes \n",
        "anndata.uns[cell_type] = contain the DE result of the corresponding cell type vs all other cells this can be used for ordering the genes by p-value, expression, and by log fold change lfc max/min/median/mean/std\n",
        "\n",
        "anndata.uns['heatmap']=dataframe with genes in index and cell types in columns containing the log10 of the scvi expression frequency for each cell type \n",
        "anndata.uns['about'] = information about the dataset\n",
        "```"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Ub-b6vZvxc53",
        "outputId": "ea9a844d-0625-4410-caf0-bbbf40e1d03e"
      },
      "source": [
        "# make one swarm df to get the shape and order of the cell types/genes to initialize adata\n",
        "mock_swarmdf=pairwise_de[pairwise_de['group1']==pairwise_de.group1[0]].pivot(values='lfc_median', columns='group2').round(2)\n",
        "\n",
        "# initialize adata and fills the X with zeroes, the lfc_median for each cell type go in the layers\n",
        "#\n",
        "swarmplot_adata=anndata.AnnData(X=np.zeros((len(mock_swarmdf.columns),len(mock_swarmdf.index))), \n",
        "                                    obs=pd.DataFrame(index=mock_swarmdf.columns),\n",
        "                                    var=pd.DataFrame(index=mock_swarmdf.index),                              \n",
        "                                )\n",
        "#loop through the celltypes and stores the lfc values for each cell in a layer\n",
        "for celltype in tqdm(mock_swarmdf.columns):\n",
        "    swarmdf=pairwise_de[pairwise_de['group1']==celltype].pivot(values='lfc_median', columns='group2').round(2)\n",
        "    swarmplot_adata.layers[celltype]=swarmdf.values.T\n",
        "\n",
        "    # now performs one vs all DE and stores those results in adata.uns for access so that the genes can be sorted according to them\n",
        "    celltype_global_de=model.differential_expression(\n",
        "        groupby='cell_type',\n",
        "        group1=celltype,\n",
        "        all_stats=False, \n",
        "        n_samples=500,\n",
        "        silent=True\n",
        "    )\n",
        "    celltype_global_de=celltype_global_de[['proba_not_de', 'bayes_factor', 'scale1', 'scale2',\n",
        "        'lfc_mean', 'lfc_median', 'lfc_std', 'lfc_min', 'lfc_max', 'comparison']]\n",
        "    celltype_global_de['log10scale1']=np.log10(celltype_global_de['scale1'])\n",
        "    celltype_global_de['log10scale2']=np.log10(celltype_global_de['scale2'])\n",
        "    swarmplot_adata.uns[celltype]=celltype_global_de \n",
        "    # also store the heatmap on the uns field for showing the mean expressison on tissue\n",
        "    # during mouseover\n",
        "    swarmplot_adata.uns['heatmap']=heatmap_df\n",
        "#rename obs and var to make clear what they hold\n",
        "swarmplot_adata.var.index.rename('gene_id',inplace=True)\n",
        "swarmplot_adata.obs.index.rename('cell_type',inplace=True)\n",
        "\n",
        "swarmplot_adata.uns['about']=\"\"\"\n",
        "bendavid2021 swarm plot data for 19 cell types and 1k genes\n",
        "layers contain the cell types data for a given swarm plot\n",
        "the uns field contain the DE result of the corresponding cell type vs all other cells\n",
        "this can be used for ordering the genes by p-value, expression, and\n",
        "by log fold change lfc max/min/median/mean/std\n",
        "this h5ad file is made to be used with WormBase wormcells-viz histogram:\n",
        "https://github.com/WormBase/wormcells-viz\n",
        "\"\"\"\n",
        "\n",
        "swarmplot_adata.write_h5ad('swarmplot_adata_bendavid2021.h5ad')\n",
        "swarmplot_adata"
      ],
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "100%|██████████| 19/19 [00:12<00:00,  1.47it/s]\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "AnnData object with n_obs × n_vars = 19 × 1094\n",
              "    uns: 'Body Wall Muscle', 'heatmap', 'Coelomocytes', 'Excretory Cells', 'Excretory Gland', 'GLR', 'Germline', 'Glia', 'Hypodermis', 'Intestine', 'Neuron', 'Pharyngeal Gland Cells', 'Pharynx and Arcade Cells', 'Seam Cells', 'Sex Myoblast', 'Somatic Gonad', 'Sphincter and Anal Muscles', 'Unknown', 'Vulval Precursor Cells', 'XXX', 'about'\n",
              "    layers: 'Body Wall Muscle', 'Coelomocytes', 'Excretory Cells', 'Excretory Gland', 'GLR', 'Germline', 'Glia', 'Hypodermis', 'Intestine', 'Neuron', 'Pharyngeal Gland Cells', 'Pharynx and Arcade Cells', 'Seam Cells', 'Sex Myoblast', 'Somatic Gonad', 'Sphincter and Anal Muscles', 'Unknown', 'Vulval Precursor Cells', 'XXX'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 15
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8Qyii5Fsz3Fl",
        "outputId": "538064d3-a27c-46e0-cebb-11b48f2dbb67"
      },
      "source": [
        "!ls -lah"
      ],
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "total 249M\n",
            "drwxr-xr-x 1 root root 4.0K Jun 10 06:01 .\n",
            "drwxr-xr-x 1 root root 4.0K Jun 10 05:51 ..\n",
            "-rw-r--r-- 1 root root 149M May 19 23:20 bendavid2021.h5ad\n",
            "-rw-r--r-- 1 root root  79M Jun 10 06:01 bendavid2021_pairwisede_19cell_types.csv\n",
            "drwxr-xr-x 4 root root 4.0K Jun  1 13:40 .config\n",
            "-rw-r--r-- 1 root root 143K Jun 10 05:55 heatmap_data_bendavid2021.h5ad\n",
            "-rw-r--r-- 1 root root  17M Jun 10 05:57 histogram_data_bendavid2021.h5ad\n",
            "drwxr-xr-x 1 root root 4.0K Jun  1 13:40 sample_data\n",
            "-rw-r--r-- 1 root root 5.1M Jun 10 06:01 swarmplot_adata_bendavid2021.h5ad\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4nGt9at-t0go"
      },
      "source": [
        ""
      ],
      "execution_count": 16,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "N94guQwGt0ZF"
      },
      "source": [
        ""
      ],
      "execution_count": 16,
      "outputs": []
    }
  ]
}