{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "wormcells_wrangle_bendavid2021_h5ad.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.7.4"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/WormBase/wormcells-notebooks/blob/main/wormcells_wrangle_bendavid2021_h5ad.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cSBTROq8_teg"
      },
      "source": [
        "## This notebook will create the `bendavid2021.h5ad` file \n",
        "\n",
        "\n",
        "**Whole-organism eQTL mapping at cellular resolution with single-cell sequencing**\n",
        "\n",
        "_Eyal Ben-David, James Boocock, Longhua Guo, Stefan Zdraljevic, Joshua S Bloom, Leonid Kruglyak_\n",
        "\n",
        "eLife 2021;10:e65857, DOI: [https://doi.org/10.7554/eLife.65857](https://doi.org/10.7554/eLife.65857)\n",
        "\n",
        "### Data description and links\n",
        "\n",
        "55,508 cells profiled with 10xv2 on L2 larvae\n",
        "\n",
        "Raw sequencing data is available at https://www.ncbi.nlm.nih.gov/bioproject/658829\n",
        "\n",
        "The authors kindly provided the original gene count matrix without the soupX modifications, which is used to create the .h5ad file in this notebook. The author provided files are:\n",
        "```\n",
        "bendavid2021_barcode_metadata_for.csv\n",
        "bendavid2021_count_matrix_wormsceQTL_unmodified.mm.gz\n",
        "bendavid2021_gene_metadata.csv\n",
        "```\n",
        "\n",
        "The R object as processed by the authors with soupX is also available on GitHub together with all the code to make the figures and calculate the numbers reported in the paper:\n",
        "\n",
        "[https://github.com/eyalbenda/worm_sceQTL/](https://github.com/eyalbenda/worm_sceQTL/)\n",
        "\n",
        "## Data wrangling conventions\n",
        "\n",
        "This document describes how WormBase should wrangle single cell RNA seq data into the [anndata](https://anndata.readthedocs.io/en/stable/) format in `.h5ad` files with standard fields, plus any number of optional fields that vary depending on the metadata the authors provide. \n",
        "\n",
        "As possible, we attempt to keep the field names lower case, short, descriptive, and only using valid Python variable names so they may be accessed via the syntax `adata.var.field_name` \n",
        "\n",
        "Our goal is to standartize the naming convention for frequently used fields so that code may be reused without headaches changing variable names. The .h5ad file should only contain genes and cells with at least one count.\n",
        "\n",
        "For now we are trying to keep these guidelines simple so that we can easily stick to them. Going forward we will update them as we learn from experience. For example, at some point WormBase may start uniformly reprocessing the data, so a version field may be needed in the adata.uns\n",
        "\n",
        "Below we provide a standard description of the mandatory fields we use in all datasets, plus some common optional ones that we have used so far (not all, each dataset may have different metadata provided by the authors). \n",
        "\n",
        "### `adata.var`: gene IDs, names and descriptions \n",
        "|Field name | Description | Type | Example value | Optionality|\n",
        "|-----------|-------------|------|-------|-----|\n",
        "| `adata.var.index` | WormBase gene ID, must be unique | string | `WBGene00010957`| Required|\n",
        "| `adata.var.gene_id` | WormBase gene ID, repeat values from index | string | `WBGene00010957`|Required\n",
        "| `adata.var.gene_name` | WormBase gene name | string | `nduo-6 `|Required|\n",
        "| `adata.var.gene_description` | WormBase short gene description. Full list available for download [here](https://www.alliancegenome.org/downloads) | string | `Predicted to have NADH dehydrogenase (ubiquinone) activity. Predicted to localize to integral component of membrane; mitochondrial membrane; and respirasome.`|Optional|\n",
        "\n",
        "### `adata.obs`: cell barcode, experiment, batch, original study, cell type\n",
        "|Field name | Description | Type | Example value | Optionality|\n",
        "|-----------|-------------|------|-------|-----|\n",
        "| `adata.var.index` | The batch name joined with cell barcode witha `+` char | string | `F4_1+TGTAACGGTTAGCTAC-1 `| Required|\n",
        "| `adata.var.study` | A unique shorthand for the study that published the data, ideally in the style <first author><year> all lower case. The .h5ad file should have the same name as the study it corresponds to.  | categorical | `taylor2020`| Required|\n",
        "| `adata.obs.sample_batch` | The run that produced the corresponding barcode. Most of the time batch and experiment will be the same, but with multiplexing sometimes an one batch can have multiple experiments | categorical | `F4_1`|Required|\n",
        "| `adata.obs.sample` | The biological sample that is in this batch | categorical string | `L2 larvae fourth repeat`|Required|\n",
        "| `adata.obs.sample_description` | Description of the sample. This is mandatory because otherwise it will be very easy to confuse two samples from their name without carefully reading the paper or contacting authors | categorical string | `F4_1`|Required|\n",
        "| `adata.obs.barcode` | The cell barcode | string | `AAACCCAAGATCGCTT-1`|Required|\n",
        "| `adata.obs.cell_type` | The cell type annotation provided by the authors. Should be `not provided` if not available | categorical | `ASJ`|Required|\n",
        "| `adata.obs.cell_subtype` | The cell subtype annotation if provided by the authors | categorical | `BWM_head_row_1`|Optional|\n",
        "| `adata.obs.tissue` | The tissue annotation if provided by the authors | categorical | `Intestine`|Optional|This document describes how WormBase should wrangle single cell RNA seq data into the [anndata](https://anndata.readthedocs.io/en/stable/) format in `.h5ad` files with standard fields, plus any number of optional fields that vary depending on the metadata the authors provide. \n",
        "\n",
        "As possible, we attempt to keep the field names lower case, short, descriptive, and only using valid Python variable names so they may be accessed via the syntax `adata.var.field_name` \n",
        "\n",
        "Our goal is to standartize the naming convention for frequently used fields so that code may be reused without headaches changing variable names. The .h5ad file should only contain genes and cells with at least one count.\n",
        "\n",
        "For now we are trying to keep these guidelines simple so that we can easily stick to them. Going forward we will update them as we learn from experience. For example, at some point WormBase may start uniformly reprocessing the data, so a version field may be needed in the adata.uns\n",
        "\n",
        "Below we provide a standard description of the mandatory fields we use in all datasets, plus some common optional ones that we have used so far (not all, each dataset may have different metadata provided by the authors). \n",
        "\n",
        "### `adata.var`: gene IDs, names and descriptions \n",
        "|Field name | Description | Type | Example value | Optionality|\n",
        "|-----------|-------------|------|-------|-----|\n",
        "| `adata.var.index` | WormBase gene ID, must be unique | string | `WBGene00010957`| Required|\n",
        "| `adata.var.gene_id` | WormBase gene ID, repeat values from index | string | `WBGene00010957`|Required\n",
        "| `adata.var.gene_name` | WormBase gene name | string | `nduo-6 `|Required|\n",
        "| `adata.var.gene_description` | WormBase short gene description. Full list available for download [here](https://www.alliancegenome.org/downloads) | string | `Predicted to have NADH dehydrogenase (ubiquinone) activity. Predicted to localize to integral component of membrane; mitochondrial membrane; and respirasome.`|Optional|\n",
        "\n",
        "### `adata.obs`: cell barcode, experiment, batch, original study, cell type\n",
        "|Field name | Description | Type | Example value | Optionality|\n",
        "|-----------|-------------|------|-------|-----|\n",
        "| `adata.var.index` | The batch name joined with cell barcode witha `+` char | string | `F4_1+TGTAACGGTTAGCTAC-1 `| Required|\n",
        "| `adata.var.study` | A unique shorthand for the study that published the data, ideally in the style <first author><year> all lower case. The .h5ad file should have the same name as the study it corresponds to.  | categorical | `taylor2020`| Required|\n",
        "| `adata.obs.batch` | The run that produced the corresponding barcode. Most of the time batch and experiment will be the same, but with multiplexing sometimes an one batch can have multiple experiments | categorical | `F4_1`|Required|\n",
        "| `adata.obs.sample` | The biological sample that is in this batch | categorical string | `L2 larvae fourth repeat`|Required|\n",
        "| `adata.obs.sample_description` | Description of the sample. This is mandatory because otherwise it will be very easy to confuse two samples from their name without carefully reading the paper or contacting authors | categorical string | `F4_1`|Required|\n",
        "| `adata.obs.barcode` | The cell barcode | string | `AAACCCAAGATCGCTT-1`|Required|\n",
        "| `adata.obs.cell_type` | The cell type annotation provided by the authors. Should be `not provided` if not available | categorical | `ASJ`|Required|\n",
        "| `adata.obs.cell_subtype` | The cell subtype annotation if provided by the authors | categorical | `BWM_head_row_1`|Optional|\n",
        "| `adata.obs.tissue` | The tissue annotation if provided by the authors | categorical | `Intestine`|Optional|"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "kLZnYKGK_IAo",
        "outputId": "e79f0450-b062-4e32-9408-81755554f037"
      },
      "source": [
        "!pip install anndata --quiet\n",
        "import anndata \n",
        "import pandas as pd\n",
        "\n",
        "anndata.__version__"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            },
            "text/plain": [
              "'0.7.6'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "56nroGB24xyi"
      },
      "source": [
        "### You need to manually download these three files from Caltech data and upload them to Colab in order to run the notebook, since Caltech data doesn't support wget\n",
        "\n",
        "```\n",
        "bendavid2021_barcode_metadata_for.csv\n",
        "bendavid2021_count_matrix_wormsceQTL_unmodified.mm.gz\n",
        "bendavid2021_gene_metadata.csv\n",
        "```\n",
        "\n",
        "https://data.caltech.edu/tindfiles/serve/fed92e1c-a935-4917-b8c6-5405d3af73f3/\n",
        "\n",
        "https://data.caltech.edu/tindfiles/serve/7a2b5d31-ebc1-4796-b350-6a8d6fc91b40/\n",
        "\n",
        "https://data.caltech.edu/tindfiles/serve/c41d4507-204f-4d1f-b9ed-ea3f9683178e/"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 680
        },
        "id": "MxJEtMPq_SmN",
        "outputId": "e2388e36-17f5-4a6e-b3da-90afcf7a6d2c"
      },
      "source": [
        "### LOAD CELL ANNOTATION DATAFRAME\n",
        "### wrangles names to conform to convention\n",
        "\n",
        "cells = pd.read_csv('bendavid2021_barcode_metadata.csv')\n",
        "display(cells.head())\n",
        "cells['barcode']=cells.barcode.str.split('_', expand=True)[2]\n",
        "cells=cells.rename(columns={'Batch':'sample_batch','neuronal_subtype':'cell_subtype'})\n",
        "cells.index=cells['sample_batch']+'+'+cells['barcode']\n",
        "cells['cell_subtype']=cells['cell_subtype'].fillna(cells['cell_type'])\n",
        "display(cells.head())\n",
        "cells['study']='bendavid2021'\n",
        "cells['sample']=cells['sample_batch']\n",
        "cells['sample_description']='C. elegans F4 segregants collected at the L2 larval stage'\n",
        "cells=cells[['study','sample_batch','sample','cell_type','cell_subtype','sample_description','barcode']]\n",
        "display(cells.head())\n",
        "\n",
        "### CHECK THAT CELLS DATAFRAME CONFORMS TO ADATA NAMING CONVENTION\n",
        "### must contain at least columns `gene_id` and `gene_name`\n",
        "for column_name in ['study','sample_batch','sample','sample_description','barcode','cell_type','cell_subtype']:\n",
        "    if column_name not in cells.columns:\n",
        "        raise ValueError(column_name + ' is not a name in gene dataframe columns')\n",
        "         \n",
        "print('cells column names look ok!')"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Batch</th>\n",
              "      <th>Size_Factor</th>\n",
              "      <th>cell_type</th>\n",
              "      <th>neuronal_subtype</th>\n",
              "      <th>total</th>\n",
              "      <th>barcode</th>\n",
              "      <th>doublet</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>102.532200</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>NaN</td>\n",
              "      <td>104863</td>\n",
              "      <td>F4_1_TGTAACGGTTAGCTAC-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>60.046012</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>NaN</td>\n",
              "      <td>61411</td>\n",
              "      <td>F4_1_GGCAGTCCAGCCTATA-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>60.384321</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>NaN</td>\n",
              "      <td>61757</td>\n",
              "      <td>F4_1_AAGTACCGTCATCCCT-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>51.692898</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>NaN</td>\n",
              "      <td>52868</td>\n",
              "      <td>F4_1_AAGATAGTCCCTCTAG-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>59.001750</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>NaN</td>\n",
              "      <td>60343</td>\n",
              "      <td>F4_1_ACCAAACCAGCTGTAT-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  Batch  Size_Factor  ...                  barcode doublet\n",
              "0  F4_1   102.532200  ...  F4_1_TGTAACGGTTAGCTAC-1   False\n",
              "1  F4_1    60.046012  ...  F4_1_GGCAGTCCAGCCTATA-1   False\n",
              "2  F4_1    60.384321  ...  F4_1_AAGTACCGTCATCCCT-1   False\n",
              "3  F4_1    51.692898  ...  F4_1_AAGATAGTCCCTCTAG-1   False\n",
              "4  F4_1    59.001750  ...  F4_1_ACCAAACCAGCTGTAT-1   False\n",
              "\n",
              "[5 rows x 7 columns]"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>sample_batch</th>\n",
              "      <th>Size_Factor</th>\n",
              "      <th>cell_type</th>\n",
              "      <th>cell_subtype</th>\n",
              "      <th>total</th>\n",
              "      <th>barcode</th>\n",
              "      <th>doublet</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>F4_1+TGTAACGGTTAGCTAC-1</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>102.532200</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>104863</td>\n",
              "      <td>TGTAACGGTTAGCTAC-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4_1+GGCAGTCCAGCCTATA-1</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>60.046012</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>61411</td>\n",
              "      <td>GGCAGTCCAGCCTATA-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4_1+AAGTACCGTCATCCCT-1</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>60.384321</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>61757</td>\n",
              "      <td>AAGTACCGTCATCCCT-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4_1+AAGATAGTCCCTCTAG-1</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>51.692898</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>52868</td>\n",
              "      <td>AAGATAGTCCCTCTAG-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4_1+ACCAAACCAGCTGTAT-1</th>\n",
              "      <td>F4_1</td>\n",
              "      <td>59.001750</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>60343</td>\n",
              "      <td>ACCAAACCAGCTGTAT-1</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                        sample_batch  Size_Factor  ...             barcode doublet\n",
              "F4_1+TGTAACGGTTAGCTAC-1         F4_1   102.532200  ...  TGTAACGGTTAGCTAC-1   False\n",
              "F4_1+GGCAGTCCAGCCTATA-1         F4_1    60.046012  ...  GGCAGTCCAGCCTATA-1   False\n",
              "F4_1+AAGTACCGTCATCCCT-1         F4_1    60.384321  ...  AAGTACCGTCATCCCT-1   False\n",
              "F4_1+AAGATAGTCCCTCTAG-1         F4_1    51.692898  ...  AAGATAGTCCCTCTAG-1   False\n",
              "F4_1+ACCAAACCAGCTGTAT-1         F4_1    59.001750  ...  ACCAAACCAGCTGTAT-1   False\n",
              "\n",
              "[5 rows x 7 columns]"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>study</th>\n",
              "      <th>sample_batch</th>\n",
              "      <th>sample</th>\n",
              "      <th>cell_type</th>\n",
              "      <th>cell_subtype</th>\n",
              "      <th>sample_description</th>\n",
              "      <th>barcode</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>F4_1+TGTAACGGTTAGCTAC-1</th>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>TGTAACGGTTAGCTAC-1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4_1+GGCAGTCCAGCCTATA-1</th>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>GGCAGTCCAGCCTATA-1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4_1+AAGTACCGTCATCCCT-1</th>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>Somatic Gonad</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>AAGTACCGTCATCCCT-1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4_1+AAGATAGTCCCTCTAG-1</th>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>Intestine</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>AAGATAGTCCCTCTAG-1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>F4_1+ACCAAACCAGCTGTAT-1</th>\n",
              "      <td>bendavid2021</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>F4_1</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>Pharynx and Arcade Cells</td>\n",
              "      <td>C. elegans F4 segregants collected at the L2 l...</td>\n",
              "      <td>ACCAAACCAGCTGTAT-1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                study  ...             barcode\n",
              "F4_1+TGTAACGGTTAGCTAC-1  bendavid2021  ...  TGTAACGGTTAGCTAC-1\n",
              "F4_1+GGCAGTCCAGCCTATA-1  bendavid2021  ...  GGCAGTCCAGCCTATA-1\n",
              "F4_1+AAGTACCGTCATCCCT-1  bendavid2021  ...  AAGTACCGTCATCCCT-1\n",
              "F4_1+AAGATAGTCCCTCTAG-1  bendavid2021  ...  AAGATAGTCCCTCTAG-1\n",
              "F4_1+ACCAAACCAGCTGTAT-1  bendavid2021  ...  ACCAAACCAGCTGTAT-1\n",
              "\n",
              "[5 rows x 7 columns]"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            "cells column names look ok!\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 654
        },
        "id": "SE7qbuzw1aSI",
        "outputId": "f3c80881-40fe-40b3-81b7-87f08a1fd7cb"
      },
      "source": [
        "### LOAD GENE DATAFRAME AND CHANGES NAMES TO CONVENTION\n",
        "genes = pd.read_csv('bendavid2021_gene_metadata.csv', index_col=0)\n",
        "display(genes.head())\n",
        "genes=genes.rename(columns={'wormbase_gene':'gene_id','gene_short_name':'gene_name'})\n",
        "genes.index.name=None\n",
        "genes=genes[['gene_id','gene_name','wbps_transcript_id', 'chromosome_name', 'start_position',\n",
        "       'end_position', 'strand', 'external_gene_id', 'external_transcript_id',\n",
        "       'wormbase_locus', 'wormbase_gseq']]\n",
        "display(genes)\n",
        "\n",
        "### CHECK THAT GENES DATAFRAME CONFORMS TO ADATA NAMING CONVENTION\n",
        "### must contain at least columns `gene_id` and `gene_name`\n",
        "for column_name in ['gene_id','gene_name']:\n",
        "    if column_name not in genes.columns:\n",
        "        raise ValueError(column_name + ' is not a name in gene dataframe columns')\n",
        "         \n",
        "print('gene column names look ok!')"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>wbps_transcript_id</th>\n",
              "      <th>chromosome_name</th>\n",
              "      <th>start_position</th>\n",
              "      <th>end_position</th>\n",
              "      <th>strand</th>\n",
              "      <th>external_gene_id</th>\n",
              "      <th>external_transcript_id</th>\n",
              "      <th>wormbase_gene</th>\n",
              "      <th>wormbase_locus</th>\n",
              "      <th>wormbase_gseq</th>\n",
              "      <th>gene_short_name</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>wbps_gene_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>WBGene00010957</th>\n",
              "      <td>MTCE.3.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>113</td>\n",
              "      <td>549</td>\n",
              "      <td>1</td>\n",
              "      <td>nduo-6</td>\n",
              "      <td>MTCE.3.1</td>\n",
              "      <td>WBGene00010957</td>\n",
              "      <td>nduo-6</td>\n",
              "      <td>MTCE.3</td>\n",
              "      <td>nduo-6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010958</th>\n",
              "      <td>MTCE.4.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>549</td>\n",
              "      <td>783</td>\n",
              "      <td>1</td>\n",
              "      <td>ndfl-4</td>\n",
              "      <td>MTCE.4.1</td>\n",
              "      <td>WBGene00010958</td>\n",
              "      <td>ndfl-4</td>\n",
              "      <td>MTCE.4</td>\n",
              "      <td>ndfl-4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010959</th>\n",
              "      <td>MTCE.11.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>1763</td>\n",
              "      <td>2635</td>\n",
              "      <td>1</td>\n",
              "      <td>nduo-1</td>\n",
              "      <td>MTCE.11.1</td>\n",
              "      <td>WBGene00010959</td>\n",
              "      <td>nduo-1</td>\n",
              "      <td>MTCE.11</td>\n",
              "      <td>nduo-1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010960</th>\n",
              "      <td>MTCE.12.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>2634</td>\n",
              "      <td>3235</td>\n",
              "      <td>1</td>\n",
              "      <td>atp-6</td>\n",
              "      <td>MTCE.12.1</td>\n",
              "      <td>WBGene00010960</td>\n",
              "      <td>atp-6</td>\n",
              "      <td>MTCE.12</td>\n",
              "      <td>atp-6</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010961</th>\n",
              "      <td>MTCE.16.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>3389</td>\n",
              "      <td>4269</td>\n",
              "      <td>1</td>\n",
              "      <td>nduo-2</td>\n",
              "      <td>MTCE.16.1</td>\n",
              "      <td>WBGene00010961</td>\n",
              "      <td>nduo-2</td>\n",
              "      <td>MTCE.16</td>\n",
              "      <td>nduo-2</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "               wbps_transcript_id chromosome_name  ...  wormbase_gseq  gene_short_name\n",
              "wbps_gene_id                                       ...                                \n",
              "WBGene00010957           MTCE.3.1           MtDNA  ...         MTCE.3           nduo-6\n",
              "WBGene00010958           MTCE.4.1           MtDNA  ...         MTCE.4           ndfl-4\n",
              "WBGene00010959          MTCE.11.1           MtDNA  ...        MTCE.11           nduo-1\n",
              "WBGene00010960          MTCE.12.1           MtDNA  ...        MTCE.12            atp-6\n",
              "WBGene00010961          MTCE.16.1           MtDNA  ...        MTCE.16           nduo-2\n",
              "\n",
              "[5 rows x 11 columns]"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "display_data",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene_id</th>\n",
              "      <th>gene_name</th>\n",
              "      <th>wbps_transcript_id</th>\n",
              "      <th>chromosome_name</th>\n",
              "      <th>start_position</th>\n",
              "      <th>end_position</th>\n",
              "      <th>strand</th>\n",
              "      <th>external_gene_id</th>\n",
              "      <th>external_transcript_id</th>\n",
              "      <th>wormbase_locus</th>\n",
              "      <th>wormbase_gseq</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>WBGene00010957</th>\n",
              "      <td>WBGene00010957</td>\n",
              "      <td>nduo-6</td>\n",
              "      <td>MTCE.3.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>113</td>\n",
              "      <td>549</td>\n",
              "      <td>1</td>\n",
              "      <td>nduo-6</td>\n",
              "      <td>MTCE.3.1</td>\n",
              "      <td>nduo-6</td>\n",
              "      <td>MTCE.3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010958</th>\n",
              "      <td>WBGene00010958</td>\n",
              "      <td>ndfl-4</td>\n",
              "      <td>MTCE.4.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>549</td>\n",
              "      <td>783</td>\n",
              "      <td>1</td>\n",
              "      <td>ndfl-4</td>\n",
              "      <td>MTCE.4.1</td>\n",
              "      <td>ndfl-4</td>\n",
              "      <td>MTCE.4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010959</th>\n",
              "      <td>WBGene00010959</td>\n",
              "      <td>nduo-1</td>\n",
              "      <td>MTCE.11.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>1763</td>\n",
              "      <td>2635</td>\n",
              "      <td>1</td>\n",
              "      <td>nduo-1</td>\n",
              "      <td>MTCE.11.1</td>\n",
              "      <td>nduo-1</td>\n",
              "      <td>MTCE.11</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010960</th>\n",
              "      <td>WBGene00010960</td>\n",
              "      <td>atp-6</td>\n",
              "      <td>MTCE.12.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>2634</td>\n",
              "      <td>3235</td>\n",
              "      <td>1</td>\n",
              "      <td>atp-6</td>\n",
              "      <td>MTCE.12.1</td>\n",
              "      <td>atp-6</td>\n",
              "      <td>MTCE.12</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00010961</th>\n",
              "      <td>WBGene00010961</td>\n",
              "      <td>nduo-2</td>\n",
              "      <td>MTCE.16.1</td>\n",
              "      <td>MtDNA</td>\n",
              "      <td>3389</td>\n",
              "      <td>4269</td>\n",
              "      <td>1</td>\n",
              "      <td>nduo-2</td>\n",
              "      <td>MTCE.16.1</td>\n",
              "      <td>nduo-2</td>\n",
              "      <td>MTCE.16</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00021597</th>\n",
              "      <td>WBGene00021597</td>\n",
              "      <td>spsb-1</td>\n",
              "      <td>Y46E12BL.4a.1</td>\n",
              "      <td>II</td>\n",
              "      <td>15207443</td>\n",
              "      <td>15229725</td>\n",
              "      <td>1</td>\n",
              "      <td>spsb-1</td>\n",
              "      <td>Y46E12BL.4a.1</td>\n",
              "      <td>spsb-1</td>\n",
              "      <td>Y46E12BL.4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00021596</th>\n",
              "      <td>WBGene00021596</td>\n",
              "      <td>spsb-2</td>\n",
              "      <td>Y46E12BL.3.1</td>\n",
              "      <td>II</td>\n",
              "      <td>15230343</td>\n",
              "      <td>15232803</td>\n",
              "      <td>1</td>\n",
              "      <td>spsb-2</td>\n",
              "      <td>Y46E12BL.3.1</td>\n",
              "      <td>spsb-2</td>\n",
              "      <td>Y46E12BL.3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00021595</th>\n",
              "      <td>WBGene00021595</td>\n",
              "      <td>Y46E12BL.2</td>\n",
              "      <td>Y46E12BL.2a.1</td>\n",
              "      <td>II</td>\n",
              "      <td>15233153</td>\n",
              "      <td>15244931</td>\n",
              "      <td>1</td>\n",
              "      <td>Y46E12BL.2</td>\n",
              "      <td>Y46E12BL.2a.1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Y46E12BL.2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00021594</th>\n",
              "      <td>WBGene00021594</td>\n",
              "      <td>tig-3</td>\n",
              "      <td>Y46E12BL.1.1</td>\n",
              "      <td>II</td>\n",
              "      <td>15244855</td>\n",
              "      <td>15251774</td>\n",
              "      <td>-1</td>\n",
              "      <td>tig-3</td>\n",
              "      <td>Y46E12BL.1.1</td>\n",
              "      <td>tig-3</td>\n",
              "      <td>Y46E12BL.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>WBGene00007064</th>\n",
              "      <td>WBGene00007064</td>\n",
              "      <td>rga-9</td>\n",
              "      <td>2RSSE.1a.1</td>\n",
              "      <td>II</td>\n",
              "      <td>15268212</td>\n",
              "      <td>15275689</td>\n",
              "      <td>1</td>\n",
              "      <td>rga-9</td>\n",
              "      <td>2RSSE.1a.1</td>\n",
              "      <td>rga-9</td>\n",
              "      <td>2RSSE.1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>20138 rows × 11 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                       gene_id   gene_name  ... wormbase_locus wormbase_gseq\n",
              "WBGene00010957  WBGene00010957      nduo-6  ...         nduo-6        MTCE.3\n",
              "WBGene00010958  WBGene00010958      ndfl-4  ...         ndfl-4        MTCE.4\n",
              "WBGene00010959  WBGene00010959      nduo-1  ...         nduo-1       MTCE.11\n",
              "WBGene00010960  WBGene00010960       atp-6  ...          atp-6       MTCE.12\n",
              "WBGene00010961  WBGene00010961      nduo-2  ...         nduo-2       MTCE.16\n",
              "...                        ...         ...  ...            ...           ...\n",
              "WBGene00021597  WBGene00021597      spsb-1  ...         spsb-1    Y46E12BL.4\n",
              "WBGene00021596  WBGene00021596      spsb-2  ...         spsb-2    Y46E12BL.3\n",
              "WBGene00021595  WBGene00021595  Y46E12BL.2  ...            NaN    Y46E12BL.2\n",
              "WBGene00021594  WBGene00021594       tig-3  ...          tig-3    Y46E12BL.1\n",
              "WBGene00007064  WBGene00007064       rga-9  ...          rga-9       2RSSE.1\n",
              "\n",
              "[20138 rows x 11 columns]"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "stream",
          "text": [
            "gene column names look ok!\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3fzM_bqUKBhX",
        "outputId": "021a3169-6977-465c-923f-ddf0025d29da"
      },
      "source": [
        "adata=anndata.read_mtx('bendavid2021_count_matrix_wormsceQTL_unmodified.mm.gz')\n",
        "adata=adata.T\n",
        "adata.var=genes\n",
        "adata.obs=cells\n",
        "\n",
        "### CHECK THAT OBS AND VAR CONFORMS TO ADATA NAMING CONVENTION\n",
        "for column_name in ['study','sample_batch','sample','sample_description','barcode','cell_type','cell_subtype']:\n",
        "    if column_name not in adata.obs.columns:\n",
        "        raise ValueError(column_name + ' is not a name in gene adata.var columns')\n",
        "for column_name in ['gene_id','gene_name']:\n",
        "    if column_name not in adata.var.columns:\n",
        "        raise ValueError(column_name + ' is not a name in gene adata.obs columns')\n",
        "         \n",
        "print('adata.var and adata.obs column names look ok!')\n",
        "adata.write_h5ad('bendavid2021.h5ad')\n",
        "adata"
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "... storing 'study' as categorical\n",
            "... storing 'sample_batch' as categorical\n",
            "... storing 'sample' as categorical\n",
            "... storing 'cell_type' as categorical\n",
            "... storing 'cell_subtype' as categorical\n",
            "... storing 'sample_description' as categorical\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "adata.var and adata.obs column names look ok!\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "... storing 'barcode' as categorical\n",
            "... storing 'chromosome_name' as categorical\n",
            "... storing 'wormbase_locus' as categorical\n",
            "... storing 'wormbase_gseq' as categorical\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "AnnData object with n_obs × n_vars = 55508 × 20138\n",
              "    obs: 'study', 'sample_batch', 'sample', 'cell_type', 'cell_subtype', 'sample_description', 'barcode'\n",
              "    var: 'gene_id', 'gene_name', 'wbps_transcript_id', 'chromosome_name', 'start_position', 'end_position', 'strand', 'external_gene_id', 'external_transcript_id', 'wormbase_locus', 'wormbase_gseq'"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 10
        }
      ]
    }
  ]
}